# Essential Gene Review

In [195]:
# Set up PATH for imports
import sys
if ".." not in sys.path:
    sys.path.append("..")
    
import cobra
import pandas as pd
from cobra.io import read_sbml_model

model = read_sbml_model("../model/Rpom_05.xml")
model

Name,Rpom_05
Memory address,12940a0d0
Number of metabolites,1797
Number of reactions,1796
Number of genes,968
Number of groups,0
Objective expression,1.0*Rpom_hwa_biomass - 1.0*Rpom_hwa_biomass_reverse_5ec2f
Compartments,"c, p, e"


In [345]:
def check_essential_rxn(rxn_id):
    essential_rxn = model.reactions.get_by_id(rxn_id)
    with model:
        model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)

        sol = model.optimize()
        print(f"Growth rate with " + rxn_id + f": {sol.objective_value:.4f} 1/hr")
        baseline_flux = sol.fluxes[essential_rxn.id]
        print(f"Flux through " + rxn_id + f": {baseline_flux}\n")

        essential_rxn.knock_out()

        sol = model.optimize()
        print(f"Growth rate without " + rxn_id + f": {sol.objective_value:.4f} 1/hr")
    if sol.objective_value == 0:
        return True, baseline_flux
    else:
        return False, baseline_flux

def get_flux(rxn_id):
    my_rxn = model.reactions.get_by_id(rxn_id)
    with model:
        model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)
        sol = model.optimize()
        print(f"Growth rate: {sol.objective_value:.4f} 1/hr")
        baseline_flux = sol.fluxes[my_rxn.id]
        print(f"Flux through {my_rxn.id}: {baseline_flux}")
        return baseline_flux
    
def add_source(rxn_id, to_add_id): 
    essential_rxn = model.reactions.get_by_id(rxn_id)
    with model:
        model.reactions.get_by_id('EX_glc').bounds = (-5.44, 0)

        # Baseline growth rate is 0, with tptase knocked out
        sol = model.optimize()
        is_essential, baseline_flux = check_essential_rxn(rxn_id)

        essential_rxn.knock_out()
        # Does adding free to_add fix things?
        source = model.add_boundary(
            model.metabolites.get_by_id(to_add_id),
            type="sink",
            reaction_id=to_add_id + "_source",
            lb=-1 * baseline_flux,  # Set to exactly the lost flux from above!
            ub=0
        )
        print(f"New reaction: {source.id}\n\t{source.reaction}\t\t{source.bounds}")

        sol = model.optimize()
        print(f"Growth rate with new reaction: {sol.objective_value:.2f} 1/hr")
        print()
        
def add_sink(rxn_id, to_sink_id): 
    essential_rxn = model.reactions.get_by_id(rxn_id)
    with model:
        model.reactions.get_by_id('EX_glc').bounds = (-5.44, 0)

        # Baseline growth rate is 0, with tptase knocked out
        sol = model.optimize()
        (is_essential, baseline_flux) = check_essential_rxn(rxn_id)

        essential_rxn.knock_out()
        # Does adding free to_add fix things?
        sink = model.add_boundary(
            model.metabolites.get_by_id(to_sink_id),
            type="sink",
            reaction_id=to_sink_id + "_sink",
            ub=baseline_flux,  # Set to exactly the lost flux from above!
            lb=0
        )
        print(f"New reaction: {sink.id}\n\t{sink.reaction}\t\t{sink.bounds}")

        sol = model.optimize()
        print(f"Growth rate with new reaction: {sol.objective_value:.2f} 1/hr")
        print()

def get_metabolite_reactions(met_id): 
    met = model.metabolites.get_by_id(met_id)

    rxns = list(met.reactions)          # list of reactions
    rxn_ids = [r.id for r in rxns]      # just IDs
    consuming = [r for r in met.reactions if r.metabolites[met] < 0]
    producing = [r for r in met.reactions if r.metabolites[met] > 0]

    print(f"Consuming reactions:")
    for rxn in consuming:
        print(f"\t{rxn.id}: {rxn.reaction}")

    print(f"Producing reactions:")
    for rxn in producing:
        print(f"\t{rxn.id}: {rxn.reaction}")
    print()
    return consuming, producing
        
        
def add_source_and_sink(rxn_id, to_add_id, to_sink_id): 
    essential_rxn = model.reactions.get_by_id(rxn_id)
    with model:
        model.reactions.get_by_id('EX_glc').bounds = (-5.44, 0)

        # Baseline growth rate is 0, with tptase knocked out
        sol = model.optimize()
        is_essential, baseline_flux = check_essential_rxn(rxn_id)

        essential_rxn.knock_out()
        # Does adding free to_add fix things?
        source = model.add_boundary(
            model.metabolites.get_by_id(to_add_id),
            type="sink",
            reaction_id=to_add_id + "_source",
            lb=-1 * baseline_flux,  # Set to exactly the lost flux from above!
            ub=0
        )
        print(f"New reaction: {source.id}\n\t{source.reaction}\t\t{source.bounds}")
        
        sink = model.add_boundary(
            model.metabolites.get_by_id(to_sink_id),
            type="sink",
            reaction_id=to_sink_id + "_sink",
            ub=baseline_flux,  # Set to exactly the lost flux from above!
            lb=0
        )
        print(f"New reaction: {sink.id}\n\t{sink.reaction}\t\t{sink.bounds}")

        sol = model.optimize()
        print(f"Growth rate with new reaction: {sol.objective_value:.2f} 1/hr")
        print()
        
def add_many_source_and_sink(rxn_id, to_add_ids, to_sink_ids): 
    essential_rxn = model.reactions.get_by_id(rxn_id)
    with model:
        model.reactions.get_by_id('EX_glc').bounds = (-5.44, 0)

        # Baseline growth rate is 0, with tptase knocked out
        sol = model.optimize()
        is_essential, baseline_flux = check_essential_rxn(rxn_id)

        essential_rxn.knock_out()
        # Does adding free to_add fix things?
        for to_add_id in to_add_ids:
            source = model.add_boundary(
                model.metabolites.get_by_id(to_add_id),
                type="sink",
                reaction_id=to_add_id + "_source",
                lb=-1 * baseline_flux,  # Set to exactly the lost flux from above!
                ub=0
            )
            print(f"New reaction: {source.id}\n\t{source.reaction}\t\t{source.bounds}")
        
        for to_sink_id in to_sink_ids:
            sink = model.add_boundary(
                model.metabolites.get_by_id(to_sink_id),
                type="sink",
                reaction_id=to_sink_id + "_sink",
                ub=baseline_flux,  # Set to exactly the lost flux from above!
                lb=0
            )
            print(f"New reaction: {sink.id}\n\t{sink.reaction}\t\t{sink.bounds}")

        sol = model.optimize()
        print(f"Growth rate with new reaction: {sol.objective_value:.2f} 1/hr")
        print()

def print_reaction_info(rxn_id):
    rxn = model.reactions.get_by_id(rxn_id)
    print(f"{rxn.id}:\n\t{rxn.reaction}\n")
    print(f"Bounds: {rxn.bounds}")
    print(f"Reversibility: {rxn.reversibility}")
    print(f"Genes: {rxn.genes}")
    print(f"Subsystem: {rxn.subsystem}")
    
def search_metabolites(words):
    cand = []
    for m in model.metabolites:
        s = (m.id + " " + m.name).lower()
        if any(k in s for k in words):
            cand.append(m)

    # Print results as aligned columns
    rows = [(m.id or "", m.name or "", m.formula or "") for m in cand]
    w_id = max([len("id"), *(len(r[0]) for r in rows)] or [2])
    w_name = max([len("name"), *(len(r[1]) for r in rows)] or [4])
    w_formula = max([len("formula"), *(len(r[2]) for r in rows)] or [7])

    print(f"{'id':<{w_id}} | {'name':<{w_name}} | {'formula':<{w_formula}}")
    print(f"{'-'*w_id}-+-{'-'*w_name}-+-{'-'*w_formula}")
    for mid, name, formula in rows:
        print(f"{mid:<{w_id}} | {name:<{w_name}} | {formula:<{w_formula}}")

## 1. ASNSYNB-RXN

In [197]:
rxn_id = "ASNSYNB-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)

ASNSYNB-RXN:
	ATP[c] + GLN[c] + L-ASPARTATE[c] + WATER[c] --> AMP[c] + ASN[c] + GLT[c] + PPI[c] + PROTON[c]

Growth rate with ASNSYNB-RXN: 0.5680 1/hr
Flux through ASNSYNB-RXN: 0.11376672577012338

Growth rate without ASNSYNB-RXN: 0.0000 1/hr


Interestingly, ASNSYNB-RXN is not in biocyc or kegg. This means it is likely not in R Pom.

Why might ASNSYNB-RXN be essential?

1. Maybe we need to generate ASN?
2. Maybe we need to eliminate L-Aspartate?



In [198]:
add_source("ASNSYNB-RXN", "ASN[c]")

Growth rate with ASNSYNB-RXN: 0.5680 1/hr
Flux through ASNSYNB-RXN: 0.11376672577012077

Growth rate without ASNSYNB-RXN: 0.0000 1/hr
New reaction: ASN[c]_source
	ASN[c] <-- 		(np.float64(-0.11376672577012077), 0)
Growth rate with new reaction: 0.57 1/hr



In [199]:
add_sink("ASNSYNB-RXN", "L-ASPARTATE[c]")

Growth rate with ASNSYNB-RXN: 0.5680 1/hr
Flux through ASNSYNB-RXN: 0.1137667257701208

Growth rate without ASNSYNB-RXN: 0.0000 1/hr
New reaction: L-ASPARTATE[c]_sink
	L-ASPARTATE[c] --> 		(0, np.float64(0.1137667257701208))
Growth rate with new reaction: 0.00 1/hr



It seems that hypothesis 1 holds.

There are 2 possibilities here:

1. Either ASN[c] is not present in R Pom, and we need to eliminate a bunch of downstream reactions...
2. R Pom has an alternative means of producing ASN[c]

Biocyc indicates ASN (L-asparagine) is indeed in R Pom, so H2 is likely true here. 

In [200]:
met_id = "ASN[c]"  
consuming, producing = get_metabolite_reactions(met_id)

Consuming reactions:
	RXN-20896: ASN[c] <=> CPD-3633[c]
	ASPARAGHYD-RXN: ASN[c] + WATER[c] --> AMMONIUM[c] + L-ASPARTATE[c]
	PROTEIN-RXN: 0.437725018 ARG[c] + 0.356722982 ASN[c] + 33.1963969 ATP[c] + 0.135523805 CYS[c] + 0.389434859 GLN[c] + 0.389434859 GLT[c] + 0.906606601 GLY[c] + 0.140197141 HIS[c] + 0.429936617 ILE[c] + 0.760177851 L-ALPHA-ALANINE[c] + 0.356722982 L-ASPARTATE[c] + 0.666714077 LEU[c] + 0.507823589 LYS[c] + 0.227430786 MET[c] + 0.274162674 PHE[c] + 0.32712617 PRO[c] + 0.319337768 SER[c] + 0.375416329 THR[c] + 0.084118581 TRP[c] + 0.204064103 TYR[c] + 0.626212319 VAL[c] --> 33.1963969 ADP[c] + PROTEIN[c] + 33.1963969 PROTON[c] + 33.1963969 Pi[c]
Producing reactions:
	ASNSYNB-RXN: ATP[c] + GLN[c] + L-ASPARTATE[c] + WATER[c] --> AMP[c] + ASN[c] + GLT[c] + PPI[c] + PROTON[c]
	RXN1RHL-12: ASN[p] + ATP[c] + WATER[c] --> ADP[c] + ASN[c] + PROTON[c] + Pi[c]



- **Reaction RXN1RHL-12 in biocyc**
- ASPARAGHYD-RXN not in biocyc
- RXN-20896 not in biocyc
- PROTEIN-RXN resembles a pseudo-reaction for protein synthesis (translation)

In [201]:
consuming, producing = get_metabolite_reactions("ASN[p]")

Consuming reactions:
	RXN1RHL-12: ASN[p] + ATP[c] + WATER[c] --> ADP[c] + ASN[c] + PROTON[c] + Pi[c]
Producing reactions:
	ASNtex: ASN[e] <=> ASN[p]



In [202]:
consuming, producing = get_metabolite_reactions("PROTEIN[c]")

Consuming reactions:
	Rpom_hwa_biomass: 31.36179486 ATP[c] + 0.001524574 COFACTORS[c] + 0.05136379 DNA[c] + 0.009825009 IONS[c] + 0.095950926 LIPID[c] + 0.026355529 MUREIN[c] + 0.000447 NADP[c] + 0.001831 NAD[c] + 0.091372857 PHB-STORAGE[c] + 0.561521914 PROTEIN[c] + 0.093597842 RNA[c] + 48.601527 WATER[c] --> 31.36179573 ADP[c] + 31.36179573 PROTON[c] + 31.35745773 Pi[c]
Producing reactions:
	PROTEIN-RXN: 0.437725018 ARG[c] + 0.356722982 ASN[c] + 33.1963969 ATP[c] + 0.135523805 CYS[c] + 0.389434859 GLN[c] + 0.389434859 GLT[c] + 0.906606601 GLY[c] + 0.140197141 HIS[c] + 0.429936617 ILE[c] + 0.760177851 L-ALPHA-ALANINE[c] + 0.356722982 L-ASPARTATE[c] + 0.666714077 LEU[c] + 0.507823589 LYS[c] + 0.227430786 MET[c] + 0.274162674 PHE[c] + 0.32712617 PRO[c] + 0.319337768 SER[c] + 0.375416329 THR[c] + 0.084118581 TRP[c] + 0.204064103 TYR[c] + 0.626212319 VAL[c] --> 33.1963969 ADP[c] + PROTEIN[c] + 33.1963969 PROTON[c] + 33.1963969 Pi[c]



Either ASN is not needed (need to modify protein synthesis reaction)

OR 

We need to give the cell ASN[e] at the start

The second option seems more plausible.

https://www.nature.com/articles/ismej20166 

States that R. pomeroyi can synthesize all amino acids except asparagine and possibly histidine.

But Wait!

https://www.pnas.org/doi/10.1073/pnas.2217200120

Shows that R. Pom can grow without asparagine in media...

So there must be a means to explain what is going on...

Likely has to do with Protein Reaction

In [203]:
check_essential_rxn("PROTEIN-RXN")

Growth rate with PROTEIN-RXN: 0.5680 1/hr
Flux through PROTEIN-RXN: 0.31892177266594507

Growth rate without PROTEIN-RXN: 0.0000 1/hr


(True, np.float64(0.31892177266594507))

What is likely going on is that either:

1. When no external asparagine is present, the cell uses reaction 6.1.1.23, using L-aspartate at asn tRNA
    -   l-asparaginyl tRNA[ASN] can  also be formed by reaction RXN-20081, 6.3.5.6-RXN, using L-aspartate-tRNA[ASN].
2. Otherwise the cell imports L-aspartate normally. 

Not in any database! Essential for translation reaction. However, translation reaction neglects tRNA. While R Pom lacks traditional ASN synthesis pathways (https://www.nature.com/articles/ismej20166 ), it is able to grow in ASN-free media (https://www.pnas.org/doi/10.1073/pnas.2217200120). This is likely because L-asparaginyl-tRNA[ASN] can be synthesized independetly of ASN, from L-Aspartate: reaction 6.1.1.23, followed by RXN-20081, 6.3.5.6-RXN. Proposed Fix: either we cut this reaction and add ASN as an essential metabolite, or we keep the reaction (which seems better to me) as a pseudo-capture of the tRNA behavior (note tRNA behavior also uses amine or ATP, so perhaps we could modify this reaction into 2 separate ones which incorporate those).

## 2. AACPS4

In [204]:
rxn_id = "AACPS4"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)

AACPS4:
	ACP[c] + ATP[c] + hdcea[c] <=> AMP[c] + PPI[c] + hdeACP[c]

Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919128167

Growth rate without AACPS4: 0.0000 1/hr


Interestingly, AACPS4 is not in biocyc or kegg. This means it is likely not in R Pom.

Why might AACPS4 be essential?

1. Maybe we need to generate hdeACP?
2. Maybe we need to eliminate hdcea?

In [205]:
add_source("AACPS4", "hdeACP[c]")
print("_____")
add_source("AACPS4", "PPI[c]")
print("_____")
add_source("AACPS4", "AMP[c]")
print("_____")

Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.0854370291910244

Growth rate without AACPS4: 0.0000 1/hr
New reaction: hdeACP[c]_source
	hdeACP[c] <-- 		(np.float64(-0.0854370291910244), 0)
Growth rate with new reaction: 0.00 1/hr

_____
Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919102389

Growth rate without AACPS4: 0.0000 1/hr
New reaction: PPI[c]_source
	PPI[c] <-- 		(np.float64(-0.08543702919102389), 0)
Growth rate with new reaction: 0.00 1/hr

_____
Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919102378

Growth rate without AACPS4: 0.0000 1/hr
New reaction: AMP[c]_source
	AMP[c] <-- 		(np.float64(-0.08543702919102378), 0)
Growth rate with new reaction: 0.00 1/hr

_____


In [206]:
add_sink("AACPS4", "hdcea[c]")
print("_____")
add_sink("AACPS4", "ACP[c]")
print("_____")
add_sink("AACPS4", "ATP[c]")
print("_____")

Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919102378

Growth rate without AACPS4: 0.0000 1/hr
New reaction: hdcea[c]_sink
	hdcea[c] --> 		(0, np.float64(0.08543702919102378))
Growth rate with new reaction: 0.00 1/hr

_____
Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919102378

Growth rate without AACPS4: 0.0000 1/hr
New reaction: ACP[c]_sink
	ACP[c] --> 		(0, np.float64(0.08543702919102378))
Growth rate with new reaction: 0.00 1/hr

_____
Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919102378

Growth rate without AACPS4: 0.0000 1/hr
New reaction: ATP[c]_sink
	ATP[c] --> 		(0, np.float64(0.08543702919102378))
Growth rate with new reaction: 0.00 1/hr

_____


In [207]:
add_source_and_sink("AACPS4", "hdeACP[c]", "hdcea[c]")


Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919102378

Growth rate without AACPS4: 0.0000 1/hr
New reaction: hdeACP[c]_source
	hdeACP[c] <-- 		(np.float64(-0.08543702919102378), 0)
New reaction: hdcea[c]_sink
	hdcea[c] --> 		(0, np.float64(0.08543702919102378))
Growth rate with new reaction: 0.00 1/hr



Confusing... Let's brute force to figure out which are necessary!

In [208]:
add_many_source_and_sink("AACPS4", ["hdeACP[c]"], ["ACP[c]"])

Growth rate with AACPS4: 0.5680 1/hr
Flux through AACPS4: 0.08543702919102303

Growth rate without AACPS4: 0.0000 1/hr
New reaction: hdeACP[c]_source
	hdeACP[c] <-- 		(np.float64(-0.08543702919102303), 0)
New reaction: ACP[c]_sink
	ACP[c] --> 		(0, np.float64(0.08543702919102303))
Growth rate with new reaction: 0.57 1/hr



So it seems like we need to generate hdeACP[c] and destroy ACP[c] somehow!



In [209]:
consuming, producing = get_metabolite_reactions("hdcea[c]")

Consuming reactions:
	AACPS4: ACP[c] + ATP[c] + hdcea[c] <=> AMP[c] + PPI[c] + hdeACP[c]
Producing reactions:
	FACOAE161: POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + WATER[c] <=> CO-A[c] + PROTON[c] + hdcea[c]



In [210]:
consuming, producing = get_metabolite_reactions("hdeACP[c]")

Consuming reactions:
	1-ACYLGLYCEROL-3-P-ACYLTRANSFER-RXN: ACYL-SN-GLYCEROL-3P[c] + hdeACP[c] --> ACP[c] + L-PHOSPHATIDATE[c]
	RXN-10462: GLYCEROL-3P[c] + hdeACP[c] --> ACP[c] + ACYL-SN-GLYCEROL-3P[c]
Producing reactions:
	AACPS4: ACP[c] + ATP[c] + hdcea[c] <=> AMP[c] + PPI[c] + hdeACP[c]



In [211]:
consuming2, producing2 = get_metabolite_reactions("ACP[c]")

Consuming reactions:
	MALONYL-COA-ACP-TRANSACYL-RXN: ACP[c] + MALONYL-COA[c] --> CO-A[c] + MALONYL-ACP[c]
	AACPS4: ACP[c] + ATP[c] + hdcea[c] <=> AMP[c] + PPI[c] + hdeACP[c]
Producing reactions:
	RXN-9523: Hexanoyl-ACPs[c] + MALONYL-ACP[c] + PROTON[c] --> 3-Oxo-octanoyl-ACPs[c] + ACP[c] + CARBON-DIOXIDE[c]
	RXN-9548: Stearoyl-ACPs[c] + WATER[c] --> ACP[c] + PROTON[c] + STEARIC_ACID[c]
	RXN-10462: GLYCEROL-3P[c] + hdeACP[c] --> ACP[c] + ACYL-SN-GLYCEROL-3P[c]
	PALMITOTRANS-RXN: KDO2-LIPID-IVA[c] + Palmitoleoyl-ACPs[c] --> ACP[c] + KDO2-PALMITOLEOYL-LIPID-IVA[c]
	RXN-10025: Octanoyl-ACPs[c] + S-ADENOSYLMETHIONINE[c] <=> 5-METHYLTHIOADENOSINE[c] + ACP[c] + CPD-10785[c] + PROTON[c]
	RXN-10022: 3-oxo-hexanoyl-ACPs[c] + S-ADENOSYLMETHIONINE[c] <=> 5-METHYLTHIOADENOSINE[c] + ACP[c] + CPD-10780[c] + PROTON[c]
	UDPNACETYLGLUCOSAMACYLTRANS-RXN: R-3-hydroxymyristoyl-ACPs[c] + UDP-N-ACETYL-D-GLUCOSAMINE[c] --> ACP[c] + UDP-OHMYR-ACETYLGLUCOSAMINE[c]
	RXN-9535: Dodecanoyl-ACPs[c] + MALONYL-ACP[c] +

In [212]:
model.metabolites.get_by_id("ACP[c]")

Metabolite identifier,ACP[c]
Name,a soluble [acyl-carrier protein]
Memory address,0x12945ac40
Formula,XH
Compartment,c
In 32 reaction(s),"RXN-9523, RXN-9548, MALONYL-COA-ACP-TRANSACYL-RXN, RXN-10462, PALMITOTRANS-RXN, RXN-10025, RXN-10022, UDPNACETYLGLUCOSAMACYLTRANS-RXN, RXN-9535, RXN3O-1803, MYRISTOYLACYLTRAN-RXN, RXN-10023,..."


In [213]:
model.metabolites.get_by_id("hdeACP[c]").name

'Cis-hexadec-9-enoyl-[acyl-carrier protein] (n-C16:1)'

Maybe necessary for RXN-10462

In [214]:
print_reaction_info("RXN-10462")
print("================")
test_essential, baseline_flux = check_essential_rxn("RXN-10462")

RXN-10462:
	GLYCEROL-3P[c] + hdeACP[c] --> ACP[c] + ACYL-SN-GLYCEROL-3P[c]

Bounds: (0.0, 1000.0)
Reversibility: False
Genes: frozenset({<Gene SPO0369 at 0x12916d8b0>})
Subsystem: 
Growth rate with RXN-10462: 0.5680 1/hr
Flux through RXN-10462: 0.04271851459543866

Growth rate without RXN-10462: 0.0000 1/hr


In [215]:
add_source("RXN-10462", "ACYL-SN-GLYCEROL-3P[c]")

Growth rate with RXN-10462: 0.5680 1/hr
Flux through RXN-10462: 0.04271851459551131

Growth rate without RXN-10462: 0.0000 1/hr
New reaction: ACYL-SN-GLYCEROL-3P[c]_source
	ACYL-SN-GLYCEROL-3P[c] <-- 		(np.float64(-0.04271851459551131), 0)
Growth rate with new reaction: 0.57 1/hr



In [ ]:
get_metabolite_reactions("ACYL-SN-GLYCEROL-3P[c]")

In [216]:
get_metabolite_reactions("ACYL-SN-GLYCEROL-3P[c]")

Consuming reactions:
	1-ACYLGLYCEROL-3-P-ACYLTRANSFER-RXN: ACYL-SN-GLYCEROL-3P[c] + hdeACP[c] --> ACP[c] + L-PHOSPHATIDATE[c]
Producing reactions:
	RXN-10462: GLYCEROL-3P[c] + hdeACP[c] --> ACP[c] + ACYL-SN-GLYCEROL-3P[c]



([<Reaction 1-ACYLGLYCEROL-3-P-ACYLTRANSFER-RXN at 0x12839ad90>],
 [<Reaction RXN-10462 at 0x13c03c9d0>])

Suspect this has something to do with RXN 10462 taking acyl-[acyl-carrier protein] as input, not hdeACP.

10462 in database:
acyl-[acyl-carrier protein] + sn-glycerol 3-phosphate → a 2-lysophosphatidate + an acyl-carrier protein

It could also be the case that there needs to be a separate hdeACP synthesis pathway or an alternative means of producing **ACYL-SN-GLYCEROL-3P[c]** which is the critical component from RXN-10462.




What if we remove RXN-10462, AACPS4?

- For hdcea use defined reaction 6.2.1.3:

a long-chain fatty acid + coenzyme A + ATP → a long-chain fatty acyl-CoA + AMP + diphosphate

We can define the reaction

**hdcea[c] + CO-A[c] + ATP[c] -> hdceCOA[c] + AMP[c] + PPI[c]**

- Next, we can replace RXN-10462 with reaction 2.3.1.15:

an acyl-CoA + sn-glycerol 3-phosphate → a 2-lysophosphatidate + coenzyme A

or 

**GLYCEROL-3P[c] + hdeCOA[c] --> CO-A[c] + ACYL-SN-GLYCEROL-3P[c]**

- Lastly, we can replace 1-ACYLGLYCEROL-3-P-ACYLTRANSFER-RXN with RXN 2.3.1.51:

an acyl-CoA + a 2-lysophosphatidate → a phosphatidate + coenzyme A

or 

**ACYL-SN-GLYCEROL-3P[c] + hdceCOA[c] -> L-PHOSPHATIDATE[c] + CO-A[c]**



## 3. FACOAE161

In [217]:
rxn_id = "FACOAE161"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)

FACOAE161:
	POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + WATER[c] <=> CO-A[c] + PROTON[c] + hdcea[c]

Growth rate with FACOAE161: 0.5680 1/hr
Flux through FACOAE161: 0.08543702919102425

Growth rate without FACOAE161: 0.0000 1/hr


In [222]:
add_source("FACOAE161", "hdcea[c]")
#add_sink("FACOAE161", "POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c]")
get_metabolite_reactions("hdcea[c]")


Growth rate with FACOAE161: 0.5680 1/hr
Flux through FACOAE161: 0.08543702919102268

Growth rate without FACOAE161: 0.0000 1/hr
New reaction: hdcea[c]_source
	hdcea[c] <-- 		(np.float64(-0.08543702919102268), 0)
Growth rate with new reaction: 0.57 1/hr

Consuming reactions:
	AACPS4: ACP[c] + ATP[c] + hdcea[c] <=> AMP[c] + PPI[c] + hdeACP[c]
Producing reactions:
	FACOAE161: POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + WATER[c] <=> CO-A[c] + PROTON[c] + hdcea[c]



([<Reaction AACPS4 at 0x129331eb0>], [<Reaction FACOAE161 at 0x1296b5f40>])

In [250]:
get_metabolite_reactions("POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c]")
get_metabolite_reactions("hdcea[c]")

Consuming reactions:
	ENOYL-COA-DELTA-ISOM-RXN-POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24//CPD0-2117.53.: POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] <=> CPD0-2117[c]
	FACOAE161: POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + WATER[c] <=> CO-A[c] + PROTON[c] + hdcea[c]
Producing reactions:

Consuming reactions:
	AACPS4: ACP[c] + ATP[c] + hdcea[c] <=> AMP[c] + PPI[c] + hdeACP[c]
Producing reactions:
	FACOAE161: POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + WATER[c] <=> CO-A[c] + PROTON[c] + hdcea[c]



([<Reaction AACPS4 at 0x129331eb0>], [<Reaction FACOAE161 at 0x1296b5f40>])

"Not in database. Required to form hdcea[c]. Reactions known to produce this compound which we could replace FACOAE161 with:

1. Reaction 3.1.1.77:

a 3-(acyloxy)acyl group of bacterial toxin + H2O → a 3-hydroxyacyl group of bacterial toxin + a fatty acid + H+

2. Reaction 1.2.1.4:

an aldehyde + NADP+ + H2O → a carboxylate + NADPH + 2 H+

Here we would have some kind of:

**hexadecenal[c] + NADP+[c] + H2O[c] -> hdcea[c] + NADPH[c] + 2H+[c]**

I cannot find hexadecenal in the model or database however...

______________

http://bigg.ucsd.edu/universal/reactions/FACOAE161

suggets that FACOAE161 is actually a reaction which produces and derives from hdcoa_c. 

Ie hdCOA[c] is equivalent to POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c].

This is not the case at the molecular formula level, so maybe we need to redefine a new hdCOA[c] species.

We can define the reaction

**hdcea[c] + CO-A[c] + ATP[c] -> POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + AMP[c] + PPI[c]**

- Next, we can replace RXN-10462 with reaction 2.3.1.15:

an acyl-CoA + sn-glycerol 3-phosphate → a 2-lysophosphatidate + coenzyme A

or 

**GLYCEROL-3P[c] + POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] --> CO-A[c] + ACYL-SN-GLYCEROL-3P[c]** (1)

- Lastly, we can replace 1-ACYLGLYCEROL-3-P-ACYLTRANSFER-RXN with RXN 2.3.1.51:

an acyl-CoA + a 2-lysophosphatidate → a phosphatidate + coenzyme A

or 

**ACYL-SN-GLYCEROL-3P[c] + POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] -> L-PHOSPHATIDATE[c] + CO-A[c]** (2)

What if we just keep (1) and (2)? And knockout both 10462 and FACOAE161?

However, maybe we need to remove POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c], and replace it with hdCOA separately. 



In [249]:
m = model.metabolites.get_by_id("POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c]")
print(m.id, m.name, m.formula)
for r in m.reactions:
    print(r.id, r.reaction)


POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24 C37H60N7O17P3S
ENOYL-COA-DELTA-ISOM-RXN-POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24//CPD0-2117.53. POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] <=> CPD0-2117[c]
FACOAE161 POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + WATER[c] <=> CO-A[c] + PROTON[c] + hdcea[c]


In [239]:
m = model.metabolites.get_by_id("CPD0-2117[c]")  # adjust formatting if needed
print(m.id, m.name, m.formula)

CPD0-2117[c] (2E)-hexadecenoyl-CoA C37H60N7O17P3S


In [240]:
m = model.metabolites.get_by_id("POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c]")  # adjust formatting if needed
print(m.id, m.name, m.formula)

POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24 C37H60N7O17P3S


In [242]:
m = model.metabolites.get_by_id("hdcea[c]")  # adjust formatting if needed
print(m.id, m.name, m.formula)

hdcea[c] palmitoleate C16H29O2


In [244]:
m = model.metabolites.get_by_id("L-PHOSPHATIDATE[c]")
print(m.id, m.name, m.formula)

L-PHOSPHATIDATE[c] a 1,2-diacyl-sn-glycerol 3-phosphate C35H63O8P


In [248]:
m = model.metabolites.get_by_id("ACYL-SN-GLYCEROL-3P[c]")
print(m.id, m.name, m.formula)

m = model.metabolites.get_by_id("GLYCEROL-3P[c]")
print(m.id, m.name, m.formula)

ACYL-SN-GLYCEROL-3P[c] a 1-acyl-sn-glycerol 3-phosphate C19H35O7P
GLYCEROL-3P[c] sn-glycerol 3-phosphate C3H7O6P


In [230]:
with model:
    model.reactions.get_by_id("FACOAE161").knock_out()
    model.reactions.get_by_id("RXN-10462").knock_out()
    
    # define relevant metabolites
    g3p = model.metabolites.get_by_id("GLYCEROL-3P[c]")
    hdcoa = model.metabolites.get_by_id("POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c]")
    coa = model.metabolites.get_by_id("CO-A[c]")
    acyl_sn_glycerol_3p = model.metabolites.get_by_id("ACYL-SN-GLYCEROL-3P[c]")
    l_phosphatidate = model.metabolites.get_by_id("L-PHOSPHATIDATE[c]")
    
    # define reaction 2.3.1.15
    rxn1 = cobra.Reaction("RXN_2.3.1.15")
    rxn1.add_metabolites({g3p: -1, hdcoa: -1, coa: 1, acyl_sn_glycerol_3p: 1})
    model.add_reactions([rxn1])
    
    # define reaction 2.3.1.51
    rxn2 = cobra.Reaction("RXN_2.3.1.51")
    rxn2.add_metabolites({acyl_sn_glycerol_3p: -1, hdcoa: -1, l_phosphatidate: 1, coa: 1})
    model.add_reactions([rxn2])
    
    # test model
    model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)
    sol = model.optimize()
    print(f"Growth rate with {sol.objective_value:.4f} 1/hr")
    


Growth rate with 0.5687 1/hr


## 4. LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.

In [231]:
rxn_id = "LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56."
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)



LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.:
	ETF-Oxidized[c] + PALMITYL-COA[c] + PROTON[c] <=> CPD0-2117[c] + ETF-Reduced[c]

Growth rate with LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.5680 1/hr
Flux through LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.08543702919102396

Growth rate without LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.0000 1/hr


In [238]:
# brute forcing until we figure out which species are key
add_source(rxn_id, "CPD0-2117[c]")
add_source(rxn_id, "ETF-Reduced[c]")
add_sink(rxn_id, "ETF-Oxidized[c]")
add_sink(rxn_id, "PALMITYL-COA[c]")
add_source_and_sink(rxn_id, "CPD0-2117[c]", "PALMITYL-COA[c]")

Growth rate with LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.5680 1/hr
Flux through LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.08543702919101856

Growth rate without LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.0000 1/hr
New reaction: CPD0-2117[c]_source
	CPD0-2117[c] <-- 		(np.float64(-0.08543702919101856), 0)
Growth rate with new reaction: 0.00 1/hr

Growth rate with LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.5680 1/hr
Flux through LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.08543702919101856

Growth rate without LONG-CHAIN-ACYL-COA-DEHYDROGENASE-RXN-PALMITYL-COA/ETF-Oxidized/PROTON//CPD0-2117/ETF-Reduced.56.: 0.0000 1/hr
New reaction: ETF-Reduced[c]_source
	ETF-Reduced[c] <-- 		(np.float64(-0.085437029

Looks like CPD0-2117[c] is required as a product and PALMITYL-COA[c] as a reactant. 

Looks like this reaction corresponds to 

Reaction 1.3.8.-

BioCyc ID: ACYLCOADEHYDROG-RXN

EC Number: 1.3.8 

https://biocyc.org/reaction?orgid=GCF_000011965&id=ACYLCOADEHYDROG-RXN

a 2,3,4-saturated fatty acyl CoA + oxidized electron-transfer flavoprotein + H+ → a (2E)-2-enoyl-CoA + reduced electron-transfer flavoprotein

However CPD0-2117 is not in the metacyc database.

In [253]:
m = model.metabolites.get_by_id("CPD0-2117[c]")
print(m.id, m.name, m.formula)

CPD0-2117[c] (2E)-hexadecenoyl-CoA C37H60N7O17P3S


CPD0-2117[c] is hdCoA!

The pathway from 2-4 are all linked and come together below

In [259]:
with model:
    model.reactions.get_by_id("FACOAE161").knock_out()
    model.reactions.get_by_id("RXN-10462").knock_out()
    
    # define relevant metabolites
    g3p = model.metabolites.get_by_id("GLYCEROL-3P[c]")
    hdcoa = model.metabolites.get_by_id("CPD0-2117[c]")
    coa = model.metabolites.get_by_id("CO-A[c]")
    acyl_sn_glycerol_3p = model.metabolites.get_by_id("ACYL-SN-GLYCEROL-3P[c]")
    l_phosphatidate = model.metabolites.get_by_id("L-PHOSPHATIDATE[c]")
    
    # define reaction 2.3.1.15
    rxn1 = cobra.Reaction("RXN_2.3.1.15")
    rxn1.add_metabolites({g3p: -1, hdcoa: -1, coa: 1, acyl_sn_glycerol_3p: 1})
    model.add_reactions([rxn1])
    
    # define reaction 2.3.1.51
    rxn2 = cobra.Reaction("RXN_2.3.1.51")
    rxn2.add_metabolites({acyl_sn_glycerol_3p: -1, hdcoa: -1, l_phosphatidate: 1, coa: 1})
    model.add_reactions([rxn2])
    
    # test model
    model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)
    sol = model.optimize()
    print(f"Growth rate with {sol.objective_value:.4f} 1/hr")
    
get_metabolite_reactions("POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c]")
    


Growth rate with 0.5687 1/hr
Consuming reactions:
	ENOYL-COA-DELTA-ISOM-RXN-POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24//CPD0-2117.53.: POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] <=> CPD0-2117[c]
	FACOAE161: POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24[c] + WATER[c] <=> CO-A[c] + PROTON[c] + hdcea[c]
Producing reactions:



([<Reaction ENOYL-COA-DELTA-ISOM-RXN-POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24//CPD0-2117.53. at 0x1296a9e20>,
  <Reaction FACOAE161 at 0x1296b5f40>],
 [])

In [260]:
with model:
    model.reactions.get_by_id("FACOAE161").knock_out()
    model.reactions.get_by_id("RXN-10462").knock_out()
    model.reactions.get_by_id("ENOYL-COA-DELTA-ISOM-RXN-POLYMER-INST-CIS-DELTA3-ENOYL-COA-C12-H24//CPD0-2117.53.").knock_out()
    
    # define relevant metabolites
    g3p = model.metabolites.get_by_id("GLYCEROL-3P[c]")
    hdcoa = model.metabolites.get_by_id("CPD0-2117[c]")
    coa = model.metabolites.get_by_id("CO-A[c]")
    acyl_sn_glycerol_3p = model.metabolites.get_by_id("ACYL-SN-GLYCEROL-3P[c]")
    l_phosphatidate = model.metabolites.get_by_id("L-PHOSPHATIDATE[c]")
    
    # define reaction 2.3.1.15
    rxn1 = cobra.Reaction("RXN_2.3.1.15")
    rxn1.add_metabolites({g3p: -1, hdcoa: -1, coa: 1, acyl_sn_glycerol_3p: 1})
    model.add_reactions([rxn1])
    
    # define reaction 2.3.1.51
    rxn2 = cobra.Reaction("RXN_2.3.1.51")
    rxn2.add_metabolites({acyl_sn_glycerol_3p: -1, hdcoa: -1, l_phosphatidate: 1, coa: 1})
    model.add_reactions([rxn2])
    
    # test model
    model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)
    sol = model.optimize()
    print(f"Growth rate with {sol.objective_value:.4f} 1/hr")

Growth rate with 0.5687 1/hr


## 5., 6., 7., 8., and 9.: CDPDIGLYSYN-RXN-2, PHOSPHASERDECARB-RXN-2, PHOSPHASERSYN-RXN-, 3.6.3.1-RXN-2, 3.6.3.1-RXN

In [261]:
rxn_id = "CDPDIGLYSYN-RXN-2"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)

CDPDIGLYSYN-RXN-2:
	CPD0-1422[c] + CTP[c] + PROTON[c] --> PPI[c] + cdpdhdecg[c]

Growth rate with CDPDIGLYSYN-RXN-2: 0.5680 1/hr
Flux through CDPDIGLYSYN-RXN-2: 0.03624377496284343

Growth rate without CDPDIGLYSYN-RXN-2: 0.0000 1/hr


In [263]:
add_source(rxn_id, "cdpdhdecg[c]")
print("_____")
add_sink(rxn_id, "CPD0-1422[c]")
print("_____")
add_source_and_sink(rxn_id, "cdpdhdecg[c]", "CPD0-1422[c]")





Growth rate with CDPDIGLYSYN-RXN-2: 0.5680 1/hr
Flux through CDPDIGLYSYN-RXN-2: 0.03624377496285229

Growth rate without CDPDIGLYSYN-RXN-2: 0.0000 1/hr
New reaction: cdpdhdecg[c]_source
	cdpdhdecg[c] <-- 		(np.float64(-0.03624377496285229), 0)
Growth rate with new reaction: 0.57 1/hr

_____
Growth rate with CDPDIGLYSYN-RXN-2: 0.5680 1/hr
Flux through CDPDIGLYSYN-RXN-2: 0.03624377496285229

Growth rate without CDPDIGLYSYN-RXN-2: 0.0000 1/hr
New reaction: CPD0-1422[c]_sink
	CPD0-1422[c] --> 		(0, np.float64(0.03624377496285229))
Growth rate with new reaction: 0.00 1/hr

_____
Growth rate with CDPDIGLYSYN-RXN-2: 0.5680 1/hr
Flux through CDPDIGLYSYN-RXN-2: 0.03624377496285229

Growth rate without CDPDIGLYSYN-RXN-2: 0.0000 1/hr
New reaction: cdpdhdecg[c]_source
	cdpdhdecg[c] <-- 		(np.float64(-0.03624377496285229), 0)
New reaction: CPD0-1422[c]_sink
	CPD0-1422[c] --> 		(0, np.float64(0.03624377496285229))
Growth rate with new reaction: 0.57 1/hr



The reaction is required as a source of cdpdhdecg[c]. Unfortunately, I can't seem to find CDPDIGLYSYN-RXN-2 in the biocyc database.

Let's investigate cdpdhdecg[c] a little more.

In [264]:
m = model.metabolites.get_by_id("cdpdhdecg[c]")
print(m.id, m.name, m.formula)

get_metabolite_reactions("cdpdhdecg[c]")




cdpdhdecg[c] CDP-1,2-dihexadecanoylglycerol C44H79N3O15P2
Consuming reactions:
	PHOSPHASERSYN-RXN-2: SER[c] + cdpdhdecg[c] --> CMP[c] + CPD-12817[c] + PROTON[c]
Producing reactions:
	CDPDIGLYSYN-RXN-2: CPD0-1422[c] + CTP[c] + PROTON[c] --> PPI[c] + cdpdhdecg[c]



([<Reaction PHOSPHASERSYN-RXN-2 at 0x12b3aeca0>],
 [<Reaction CDPDIGLYSYN-RXN-2 at 0x12966c820>])

I can't find cdpdhdecg[c] in the database either. Why is it required here? apparently for PHOSPHASERSYN-RXN-2.

In [268]:
rxn_id = "PHOSPHASERSYN-RXN-2"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)
print("_____")
add_source(rxn_id, "CPD-12817[c]")

PHOSPHASERSYN-RXN-2:
	SER[c] + cdpdhdecg[c] --> CMP[c] + CPD-12817[c] + PROTON[c]

Growth rate with PHOSPHASERSYN-RXN-2: 0.5680 1/hr
Flux through PHOSPHASERSYN-RXN-2: 0.036243774962852394

Growth rate without PHOSPHASERSYN-RXN-2: 0.0000 1/hr
_____
Growth rate with PHOSPHASERSYN-RXN-2: 0.5680 1/hr
Flux through PHOSPHASERSYN-RXN-2: 0.03624377496285228

Growth rate without PHOSPHASERSYN-RXN-2: 0.0000 1/hr
New reaction: CPD-12817[c]_source
	CPD-12817[c] <-- 		(np.float64(-0.03624377496285228), 0)
Growth rate with new reaction: 0.57 1/hr



In [269]:
m = model.metabolites.get_by_id("CPD-12817[c]")
print(m.id, m.name, m.formula)


CPD-12817[c] 1,2-dipalmitoyl-phosphatidylserine C38H73NO10P


PHOSPHASERSYN-RXN-2 is required for production of CPD-12817[c]. 

Once again, this does not seem to be in the database.

In [270]:
get_metabolite_reactions("CPD-12817[c]")

Consuming reactions:
	PHOSPHASERDECARB-RXN-2: CPD-12817[c] + PROTON[c] --> CARBON-DIOXIDE[c] + CPD-12819[c]
Producing reactions:
	PHOSPHASERSYN-RXN-2: SER[c] + cdpdhdecg[c] --> CMP[c] + CPD-12817[c] + PROTON[c]



([<Reaction PHOSPHASERDECARB-RXN-2 at 0x12b3bc970>],
 [<Reaction PHOSPHASERSYN-RXN-2 at 0x12b3aeca0>])

So CPD-12817[c] is needed for reaction PHOSPHASERDECARB-RXN-2.

In [271]:
rxn_id = "PHOSPHASERDECARB-RXN-2"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)
print("_____")
add_source(rxn_id, "CPD-12819[c]")

PHOSPHASERDECARB-RXN-2:
	CPD-12817[c] + PROTON[c] --> CARBON-DIOXIDE[c] + CPD-12819[c]

Growth rate with PHOSPHASERDECARB-RXN-2: 0.5680 1/hr
Flux through PHOSPHASERDECARB-RXN-2: 0.036243774962852394

Growth rate without PHOSPHASERDECARB-RXN-2: 0.0000 1/hr
_____
Growth rate with PHOSPHASERDECARB-RXN-2: 0.5680 1/hr
Flux through PHOSPHASERDECARB-RXN-2: 0.03624377496285228

Growth rate without PHOSPHASERDECARB-RXN-2: 0.0000 1/hr
New reaction: CPD-12819[c]_source
	CPD-12819[c] <-- 		(np.float64(-0.03624377496285228), 0)
Growth rate with new reaction: 0.57 1/hr



In [275]:
m = model.metabolites.get_by_id("CPD-12819[c]")
print(m.id, m.name, m.formula)
get_metabolite_reactions("CPD-12819[c]")

CPD-12819[c] [] C37H74N1O8P1
Consuming reactions:
	LIPID-RXN: 0.186220192 CPD-12819[c] + 0.478848946 CPD-12819[p] + 0.564392674 CPD-17086[c] + 0.219487198 CPD-17086[p] + 0.000573 UNDECAPRENYL-DIPHOSPHATE[c] --> LIPID[c]
	3.6.3.1-RXN-2: ATP[c] + CPD-12819[c] + WATER[c] --> ADP[c] + CPD-12819[p] + PROTON[c] + Pi[c]
Producing reactions:
	PHOSPHASERDECARB-RXN-2: CPD-12817[c] + PROTON[c] --> CARBON-DIOXIDE[c] + CPD-12819[c]



([<Reaction LIPID-RXN at 0x13c40ed00>,
  <Reaction 3.6.3.1-RXN-2 at 0x12930b250>],
 [<Reaction PHOSPHASERDECARB-RXN-2 at 0x12b3bc970>])

PHOSPHASERDECARB-RXN-2 is required for the production of CPD-12819[c]. And once again, I cannot find CPD-12819[c] in the database.

In [278]:
rxn_id = "LIPID-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)
print("_____")
add_source(rxn_id, "LIPID[c]")

LIPID-RXN:
	0.186220192 CPD-12819[c] + 0.478848946 CPD-12819[p] + 0.564392674 CPD-17086[c] + 0.219487198 CPD-17086[p] + 0.000573 UNDECAPRENYL-DIPHOSPHATE[c] --> LIPID[c]

Growth rate with LIPID-RXN: 0.5680 1/hr
Flux through LIPID-RXN: 0.05449625143010645

Growth rate without LIPID-RXN: 0.0000 1/hr
_____
Growth rate with LIPID-RXN: 0.5680 1/hr
Flux through LIPID-RXN: 0.05449625143010664

Growth rate without LIPID-RXN: 0.0000 1/hr
New reaction: LIPID[c]_source
	LIPID[c] <-- 		(np.float64(-0.05449625143010664), 0)
Growth rate with new reaction: 0.57 1/hr



In [283]:
rxn_id = "3.6.3.1-RXN-2"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)
print("_____")
add_source(rxn_id, "CPD-12819[p]")
print("_____")
get_metabolite_reactions("CPD-12819[p]")

3.6.3.1-RXN-2:
	ATP[c] + CPD-12819[c] + WATER[c] --> ADP[c] + CPD-12819[p] + PROTON[c] + Pi[c]

Growth rate with 3.6.3.1-RXN-2: 0.5680 1/hr
Flux through 3.6.3.1-RXN-2: 0.02609547255825798

Growth rate without 3.6.3.1-RXN-2: 0.0000 1/hr
_____
Growth rate with 3.6.3.1-RXN-2: 0.5680 1/hr
Flux through 3.6.3.1-RXN-2: 0.026095472558257556

Growth rate without 3.6.3.1-RXN-2: 0.0000 1/hr
New reaction: CPD-12819[p]_source
	CPD-12819[p] <-- 		(np.float64(-0.026095472558257556), 0)
Growth rate with new reaction: 0.57 1/hr

_____
Consuming reactions:
	LIPID-RXN: 0.186220192 CPD-12819[c] + 0.478848946 CPD-12819[p] + 0.564392674 CPD-17086[c] + 0.219487198 CPD-17086[p] + 0.000573 UNDECAPRENYL-DIPHOSPHATE[c] --> LIPID[c]
Producing reactions:
	3.6.3.1-RXN-2: ATP[c] + CPD-12819[c] + WATER[c] --> ADP[c] + CPD-12819[p] + PROTON[c] + Pi[c]



([<Reaction LIPID-RXN at 0x13c40ed00>],
 [<Reaction 3.6.3.1-RXN-2 at 0x12930b250>])

CPD-12819 seems required for LIPID manufacturing, which also requires that it gets moved to periplasm from cytoplasm.  

CPD-17086 is also required for LIPID manufacturing. 

Neither CPD-17086 nor CPD-12819 are in the biocyc database for R Pom.

We probably need to rethink how the LIPID biomass calculation is done. 

In [286]:
m = model.metabolites.get_by_id("CPD-12819[c]")
print(m.id, m.name, m.formula)

m = model.metabolites.get_by_id("CPD-17086[c]")
print(m.id, m.name, m.formula)

CPD-12819[c] [] C37H74N1O8P1
CPD-17086[c] [] C37H70N1O8P1


In [285]:
rxn_id = "3.6.3.1-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)

3.6.3.1-RXN:
	ATP[c] + CPD-17086[c] + WATER[c] --> ADP[c] + CPD-17086[p] + PROTON[c] + Pi[c]

Growth rate with 3.6.3.1-RXN: 0.5680 1/hr
Flux through 3.6.3.1-RXN: 0.011961229527897792

Growth rate without 3.6.3.1-RXN: 0.0000 1/hr


Both 17086, 12819 are phosphatidylethanolamines

There might be a way to model general PE formation from phosphatidylserines (see Ruegeria pomeroyi DSS-3 (GCF_000011965.2) EC 4.1.1.65 -- phosphatidylserine decarboxylase)

PEs are known to exist in R Pom with 36 carbon atoms. (See Phosphatidylethanolamines-36-2)


## 10. GLYCOLALD-DEHYDROG-RXN

In [287]:
rxn_id = "GLYCOLALD-DEHYDROG-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)

GLYCOLALD-DEHYDROG-RXN:
	GLYCOLALDEHYDE[c] + NAD[c] + WATER[c] --> GLYCOLATE[c] + NADH[c] + 2.0 PROTON[c]

Growth rate with GLYCOLALD-DEHYDROG-RXN: 0.5680 1/hr
Flux through GLYCOLALD-DEHYDROG-RXN: 0.0003799649647626406

Growth rate without GLYCOLALD-DEHYDROG-RXN: 0.0000 1/hr


In [289]:
add_source(rxn_id, "GLYCOLATE[c]")
add_sink(rxn_id, "GLYCOLALDEHYDE[c]")

Growth rate with GLYCOLALD-DEHYDROG-RXN: 0.5680 1/hr
Flux through GLYCOLALD-DEHYDROG-RXN: 0.00037996496476265247

Growth rate without GLYCOLALD-DEHYDROG-RXN: 0.0000 1/hr
New reaction: GLYCOLATE[c]_source
	GLYCOLATE[c] <-- 		(np.float64(-0.00037996496476265247), 0)
Growth rate with new reaction: 0.00 1/hr

Growth rate with GLYCOLALD-DEHYDROG-RXN: 0.5680 1/hr
Flux through GLYCOLALD-DEHYDROG-RXN: 0.00037996496476265247

Growth rate without GLYCOLALD-DEHYDROG-RXN: 0.0000 1/hr
New reaction: GLYCOLALDEHYDE[c]_sink
	GLYCOLALDEHYDE[c] --> 		(0, np.float64(0.00037996496476265247))
Growth rate with new reaction: 0.57 1/hr



Not in database. This reaction is required as a sink for GLYCOLALDEHYDE[c].

According to biocyc, GLYCOLALDEHYDE is only consumed via reaction 1.1.1.2 or 1.2.1.4

- reaction 1.1.1.2:

an aldehyde + NADPH + H+ -> an alcohol + NADP+

So we could implement the similar reaction:

GLYCOLALDEHYDE[c] + NADPH[c] + PROTON[c] -> ethylene-glycol[c] + NADP+[c]

But ethylene glycol[c] is not in the database...

- reaction 1.2.1.4:

an aldehyde + NADP+ + H2O → a carboxylate + NADPH + 2 H+

So we could implement the similar reaction:

GLYCOLALDEHYDE[c] + NADP[c] +  H2O[c] -> glycolate[c] + NADPH[c] + 2 PROTON[c]

This is basically the same reaction as GLYCOLALD-DEHYDROG-RXN but at least in the database. It will also be essential as a sink for GLYCOLALDEHYDE.


In [303]:
m = model.metabolites.get_by_id("GLYCOLALDEHYDE[c]")
print(m.id, m.name, m.formula)

get_metabolite_reactions("GLYCOLALDEHYDE[c]")

get_metabolite_reactions("GLYCOLATE[c]")


GLYCOLALDEHYDE[c] glycolaldehyde C2H4O2
Consuming reactions:
	GLYCOLALD-DEHYDROG-RXN: GLYCOLALDEHYDE[c] + NAD[c] + WATER[c] --> GLYCOLATE[c] + NADH[c] + 2.0 PROTON[c]
Producing reactions:
	H2NEOPTERINALDOL-RXN: DIHYDRO-NEO-PTERIN[c] --> AMINO-OH-HYDROXYMETHYL-DIHYDROPTERIDINE[c] + GLYCOLALDEHYDE[c]
	RXN-10857: CPD-11770[c] --> AMINO-OH-HYDROXYMETHYL-DIHYDROPTERIDINE[c] + GLYCOLALDEHYDE[c]
	RXN0-6563: CPD0-2189[c] <=> GLYCOLALDEHYDE[c] + GLY[c]

Consuming reactions:
Producing reactions:
	GLYCOLATEtpp: GLYCOLATE[p] <=> GLYCOLATE[c]
	GLYCOLATEDEHYDRO-RXN: GLYOX[c] + NADH[c] + PROTON[c] <-- GLYCOLATE[c] + NAD[c]
	GLYCOLALD-DEHYDROG-RXN: GLYCOLALDEHYDE[c] + NAD[c] + WATER[c] --> GLYCOLATE[c] + NADH[c] + 2.0 PROTON[c]
	AMIDASE-RXN-CPD-14534/WATER//GLYCOLLATE/AMMONIUM.37.: CPD-14534[c] + WATER[c] --> AMMONIUM[c] + GLYCOLATE[c]
	GPH-RXN: CPD-67[c] + WATER[c] --> GLYCOLATE[c] + Pi[c]



([],
 [<Reaction GLYCOLATEtpp at 0x1296e7910>,
  <Reaction GLYCOLATEDEHYDRO-RXN at 0x1296e7520>,
  <Reaction GLYCOLALD-DEHYDROG-RXN at 0x1296eea30>,
  <Reaction AMIDASE-RXN-CPD-14534/WATER//GLYCOLLATE/AMMONIUM.37. at 0x129646580>,
  <Reaction GPH-RXN at 0x1296f5fd0>])

In [300]:
search_metabolites(["glycol","C2H6O2"])



id                               | name                                    | formula        
---------------------------------+-----------------------------------------+----------------
CPD-14534[c]                     | glycolamide                             | C2H5NO2        
GLYCOLATE[c]                     | glycolate                               | C2H3O3         
GLYCOLATE[e]                     | glycolate                               | C2H3O3         
GLYCOLALDEHYDE[c]                | glycolaldehyde                          | C2H4O2         
GLYCOLATE[p]                     | glycolate                               | C2H3O3         
CPD-67[c]                        | 2-phosphoglycolate                      | C2H2O6P        
CPD-11878[c]                     | 3,4-dihydroxyphenylglycol               | C8H10O4        
DIHYDROXYPHENYLGLYCOLALDEHYDE[c] | 3,4-dihydroxyphenylglycolaldehyde       | C8H8O4         
CPD-11497[c]                     | 3-methoxy-4-hydroxyphenylglycol    

In [ ]:
with model:
    model.reactions.get_by_id("GLYCOLALD-DEHYDROG-RXN").knock_out()
    
    # define relevant metabolites
    NADP = model.metabolites.get_by_id("NADP[c]")
    NADPH = model.metabolites.get_by_id("NADPH[c]")
    PROTON = model.metabolites.get_by_id("PROTON[c]")
    WATER = model.metabolites.get_by_id("WATER[c]")
    glycolaldehyde = model.metabolites.get_by_id("GLYCOLALDEHYDE[c]")
    glycolate = model.metabolites.get_by_id("GLYCOLATE[c]")
    
    # define reaction 1.2.1.4
    rxn1 = cobra.Reaction("RXN_1.2.1.4")
    #GLYCOLALDEHYDE[c] + NADP[c] +  H2O[c] -> glycolate[c] + NADPH[c] + 2 PROTON[c]
    rxn1.add_metabolites({glycolaldehyde: -1, NADP: -1, WATER: -1, glycolate: 1, NADPH: 1, PROTON: 2})
    model.add_reactions([rxn1])
    
    # test model
    model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)
    sol = model.optimize()
    print(f"Growth rate with {sol.objective_value:.4f} 1/hr")

Growth rate with 0.5680 1/hr


## 11: RIBOPHOSPHAT-RXN

In [315]:
rxn_id = "RIBOPHOSPHAT-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)

print("_____")
m = model.metabolites.get_by_id("AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE[c]")
print(m.id, m.name, m.formula)
print("_____")
m = model.metabolites.get_by_id("CPD-1086[c]")
print(m.id, m.name, m.formula)


RIBOPHOSPHAT-RXN:
	CPD-1086[c] + WATER[c] --> AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE[c] + Pi[c]

Growth rate with RIBOPHOSPHAT-RXN: 0.5680 1/hr
Flux through RIBOPHOSPHAT-RXN: 0.0002533099765084392

Growth rate without RIBOPHOSPHAT-RXN: 0.0000 1/hr
_____
AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE[c] 5-amino-6-(D-ribitylamino)uracil C9H16N4O6
_____
CPD-1086[c] 5-amino-6-(5-phospho-D-ribitylamino)uracil C9H15N4O9P


In [313]:
add_sink(rxn_id, "CPD-1086[c]")
add_source(rxn_id, "AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE[c]")

Growth rate with RIBOPHOSPHAT-RXN: 0.5680 1/hr
Flux through RIBOPHOSPHAT-RXN: 0.0002533099765084352

Growth rate without RIBOPHOSPHAT-RXN: 0.0000 1/hr
New reaction: CPD-1086[c]_sink
	CPD-1086[c] --> 		(0, np.float64(0.0002533099765084352))
Growth rate with new reaction: 0.00 1/hr

Growth rate with RIBOPHOSPHAT-RXN: 0.5680 1/hr
Flux through RIBOPHOSPHAT-RXN: 0.00025330997650843744

Growth rate without RIBOPHOSPHAT-RXN: 0.0000 1/hr
New reaction: AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE[c]_source
	AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE[c] <-- 		(np.float64(-0.00025330997650843744), 0)
Growth rate with new reaction: 0.57 1/hr



RIBOPHOSPHAT-RXN is required for production of AMINO-RIBOSYLAMINO-1H-3H-PYR-DIONE[c].

RIBOPHOSPHAT-RXN is indeed in the database!
5-amino-6-(5-phospho-D-ribitylamino)uracil + H2O → 5-amino-6-(D-ribitylamino)uracil + phosphate

RIBOPHOSPHAT-RXN is a key step in the riboflavin biosynthesis pathway in bacteria. This pathway is essential for bacteria unless riboflavin is imported separately. Essential because riboflavin-> FMN -> FAD and FMN and FAD are required for multiple redox bacterial reactions. 

No change is needed here I think.






In [316]:
search_metabolites(["riboflavin"])

id            | name               | formula   
--------------+--------------------+-----------
RIBOFLAVIN[c] | riboflavin         | C17H19N4O6
CPD-316[c]    | reduced riboflavin | C17H22N4O6


In [317]:
add_source(rxn_id, "RIBOFLAVIN[c]")

Growth rate with RIBOPHOSPHAT-RXN: 0.5680 1/hr
Flux through RIBOPHOSPHAT-RXN: 0.00025330997650843744

Growth rate without RIBOPHOSPHAT-RXN: 0.0000 1/hr
New reaction: RIBOFLAVIN[c]_source
	RIBOFLAVIN[c] <-- 		(np.float64(-0.00025330997650843744), 0)
Growth rate with new reaction: 0.57 1/hr



Indeed, it all has to do with riboflavin!

## 12: 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN

In [318]:
rxn_id = "3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)


3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN:
	3-OCTAPRENYL-4-HYDROXYBENZOATE[c] + PROTON[c] --> 2-OCTAPRENYLPHENOL[c] + CARBON-DIOXIDE[c]

Growth rate with 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.5680 1/hr
Flux through 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.00012665498825421931

Growth rate without 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.0000 1/hr


In [325]:
add_source(rxn_id, "2-OCTAPRENYLPHENOL[c]")

m = model.metabolites.get_by_id("2-OCTAPRENYLPHENOL[c]")
print(m.id, m.name, m.formula)


Growth rate with 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.5680 1/hr
Flux through 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.00012665498825421747

Growth rate without 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.0000 1/hr
New reaction: 2-OCTAPRENYLPHENOL[c]_source
	2-OCTAPRENYLPHENOL[c] <-- 		(np.float64(-0.00012665498825421747), 0)
Growth rate with new reaction: 0.57 1/hr

2-OCTAPRENYLPHENOL[c] 2-octaprenylphenol C46H70O


This reaction is necessary for the production of 2-OCTAPRENYLPHENOL[c].

This reaction is in the database.

This reaction and 2-OCTAPRENYLPHENOL are necessary in the ubiquinol-8 biosynthesis pathway (production of ubiquinol-8).

This is necessary because ubiquinol transports electrons from Complexes I and II to Complex III in the electron transport chain; and is therefore needed for oxidative phosphorylation and ATP generation.

However, in our model, this reaction is necessary because it feeds the COFACTORS[c] generation reaction. 

COFACTORS is necessary for:

	Rpom_hwa_biomass: 31.36179486 ATP[c] + 0.001524574 COFACTORS[c] + 0.05136379 DNA[c] + 0.009825009 IONS[c] + 0.095950926 LIPID[c] + 0.026355529 MUREIN[c] + 0.000447 NADP[c] + 0.001831 NAD[c] + 0.091372857 PHB-STORAGE[c] + 0.561521914 PROTEIN[c] + 0.093597842 RNA[c] + 48.601527 WATER[c] --> 31.36179573 ADP[c] + 31.36179573 PROTON[c] + 31.35745773 Pi[c] 

(ie for growth directly)

Ubiquinol-8 is generated separately as a pure source it seems. Perhaps we should implement the full ubiquinol-8 biosynthesis pathway in the model for greater accuracy? Ubiquinol is only modeled as produced, but never as consumed...



In [332]:
m = model.metabolites.get_by_id("2-OCTAPRENYL-6-HYDROXYPHENOL[c]")
get_metabolite_reactions(m.id)
print(m.id, m.name, m.formula)

Consuming reactions:
	COFACTOR-RXN: 0.14627036 2-OCTAPRENYL-6-HYDROXYPHENOL[c] + 0.377810437 CO-A[c] + 0.14627036 FAD[c] + 0.14627036 FORMYL-THF-GLU-N[c] + 0.14627036 METHYLENE-THF-GLU-N[c] + 0.14627036 PROTOHEME[c] + 0.14627036 PYRIDOXAL_PHOSPHATE[c] + 0.14627036 RIBOFLAVIN[c] + 0.14627036 S-ADENOSYLMETHIONINE[c] + 0.14627036 THF-GLU-N[c] + 0.14627036 THIAMINE-PYROPHOSPHATE[c] --> COFACTORS[c]
	2-OCTAPRENYL-6-OHPHENOL-METHY-RXN: 2-OCTAPRENYL-6-HYDROXYPHENOL[c] + S-ADENOSYLMETHIONINE[c] --> 2-OCTAPRENYL-6-METHOXYPHENOL[c] + ADENOSYL-HOMO-CYS[c] + PROTON[c]
Producing reactions:
	2-OCTAPRENYLPHENOL-HYDROX-RXN: 2-OCTAPRENYLPHENOL[c] + NADPH[c] + OXYGEN-MOLECULE[c] + PROTON[c] --> 2-OCTAPRENYL-6-HYDROXYPHENOL[c] + NADP[c] + WATER[c]

2-OCTAPRENYL-6-HYDROXYPHENOL[c] 3-(all-trans-octaprenyl)benzene-1,2-diol C46H70O2


In [468]:
search_metabolites(["ubiquinol"])

id                                    | name                         | formula 
--------------------------------------+------------------------------+---------
CPD-9958[c]                           | ubiquinol-10                 | C59H92O4
UBIQUINOL-30[c]                       | ubiquinol-6                  | C39H60O4
CPD-9956[c]                           | ubiquinol-8                  | C49H76O4
OCTAPRENYL-METHYL-OH-METHOXY-BENZQ[c] | 3-demethylubiquinol-8        | C48H74O4
CPD-9873[c]                           | 3-demethylubiquinol-10       | C58H90O4
CPD-9861[c]                           | 3-demethylubiquinol-<i>7</i> | C43H66O4
CPD-9955[c]                           | ubiquinol-<i>7</i>           | C44H68O4


In [333]:
m = model.metabolites.get_by_id("COFACTORS[c]")
get_metabolite_reactions(m.id)
print(m.id, m.name, m.formula)


Consuming reactions:
	Rpom_hwa_biomass: 31.36179486 ATP[c] + 0.001524574 COFACTORS[c] + 0.05136379 DNA[c] + 0.009825009 IONS[c] + 0.095950926 LIPID[c] + 0.026355529 MUREIN[c] + 0.000447 NADP[c] + 0.001831 NAD[c] + 0.091372857 PHB-STORAGE[c] + 0.561521914 PROTEIN[c] + 0.093597842 RNA[c] + 48.601527 WATER[c] --> 31.36179573 ADP[c] + 31.36179573 PROTON[c] + 31.35745773 Pi[c]
Producing reactions:
	COFACTOR-RXN: 0.14627036 2-OCTAPRENYL-6-HYDROXYPHENOL[c] + 0.377810437 CO-A[c] + 0.14627036 FAD[c] + 0.14627036 FORMYL-THF-GLU-N[c] + 0.14627036 METHYLENE-THF-GLU-N[c] + 0.14627036 PROTOHEME[c] + 0.14627036 PYRIDOXAL_PHOSPHATE[c] + 0.14627036 RIBOFLAVIN[c] + 0.14627036 S-ADENOSYLMETHIONINE[c] + 0.14627036 THF-GLU-N[c] + 0.14627036 THIAMINE-PYROPHOSPHATE[c] --> COFACTORS[c]

COFACTORS[c]  None


In [331]:
m = model.metabolites.get_by_id("2-OCTAPRENYL-6-HYDROXYPHENOL[c]")
print(m.id, m.name, m.formula)


2-OCTAPRENYL-6-HYDROXYPHENOL[c] 3-(all-trans-octaprenyl)benzene-1,2-diol C46H70O2


In [322]:
search_metabolites(["ubiquinol"])

id                                    | name                         | formula 
--------------------------------------+------------------------------+---------
CPD-9958[c]                           | ubiquinol-10                 | C59H92O4
UBIQUINOL-30[c]                       | ubiquinol-6                  | C39H60O4
CPD-9956[c]                           | ubiquinol-8                  | C49H76O4
OCTAPRENYL-METHYL-OH-METHOXY-BENZQ[c] | 3-demethylubiquinol-8        | C48H74O4
CPD-9873[c]                           | 3-demethylubiquinol-10       | C58H90O4
CPD-9861[c]                           | 3-demethylubiquinol-<i>7</i> | C43H66O4
CPD-9955[c]                           | ubiquinol-<i>7</i>           | C44H68O4


In [335]:
add_source(rxn_id, "COFACTORS[c]")

Growth rate with 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.5680 1/hr
Flux through 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.00012665498825421747

Growth rate without 3-OCTAPRENYL-4-OHBENZOATE-DECARBOX-RXN: 0.0000 1/hr
New reaction: COFACTORS[c]_source
	COFACTORS[c] <-- 		(np.float64(-0.00012665498825421747), 0)
Growth rate with new reaction: 0.08 1/hr



In [336]:
m = model.metabolites.get_by_id("CPD-9956[c]")
print(m.id, m.name, m.formula)

get_metabolite_reactions("CPD-9956[c]")





CPD-9956[c] ubiquinol-8 C49H76O4
Consuming reactions:
Producing reactions:
	GLYC3PDEHYDROG-RXN: GLYCEROL-3P[c] + UBIQUINONE-8[c] --> CPD-9956[c] + DIHYDROXY-ACETONE-PHOSPHATE[c]



([], [<Reaction GLYC3PDEHYDROG-RXN at 0x13c76ebe0>])

## 13. ABC-32-RXN

In [337]:
rxn_id = "ABC-32-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)




ABC-32-RXN:
	ATP[c] + THIAMINE[p] + WATER[c] --> ADP[c] + PROTON[c] + Pi[c] + THIAMINE[c]

Growth rate with ABC-32-RXN: 0.5680 1/hr
Flux through ABC-32-RXN: 0.00012665498825421942

Growth rate without ABC-32-RXN: 0.0000 1/hr


In [339]:
add_sink(rxn_id, "THIAMINE[p]")
add_source(rxn_id, "THIAMINE[c]")

Growth rate with ABC-32-RXN: 0.5680 1/hr
Flux through ABC-32-RXN: 0.00012665498825421742

Growth rate without ABC-32-RXN: 0.0000 1/hr
New reaction: THIAMINE[p]_sink
	THIAMINE[p] --> 		(0, np.float64(0.00012665498825421742))
Growth rate with new reaction: 0.00 1/hr

Growth rate with ABC-32-RXN: 0.5680 1/hr
Flux through ABC-32-RXN: 0.00012665498825421742

Growth rate without ABC-32-RXN: 0.0000 1/hr
New reaction: THIAMINE[c]_source
	THIAMINE[c] <-- 		(np.float64(-0.00012665498825421742), 0)
Growth rate with new reaction: 0.57 1/hr



ABC-32-RXN is not in the database. However, it is required in our model for the production of THIAMINE[c].

Thiamine is produced by reaction 3.1.3.100 

thiamine phosphate + H2O → thiamine + phosphate

This is in the model but has 0 flux (see below):

In [348]:
m = model.metabolites.get_by_id("THIAMINE[c]")
print(m.id, m.name, m.formula)

get_metabolite_reactions("THIAMINE[c]")


THIAMINE[c] thiamine C12H17N4OS
Consuming reactions:
	THIAMIN-PYROPHOSPHOKINASE-RXN: ATP[c] + THIAMINE[c] --> AMP[c] + PROTON[c] + THIAMINE-PYROPHOSPHATE[c]
	THIAMINASE-RXN: THIAMINE[c] + WATER[c] --> HMP[c] + PROTON[c] + THZ[c]
Producing reactions:
	ABC-32-RXN: ATP[c] + THIAMINE[p] + WATER[c] --> ADP[c] + PROTON[c] + Pi[c] + THIAMINE[c]
	RXNQT-4191: THIAMINE-P[c] + WATER[c] --> Pi[c] + THIAMINE[c]



([<Reaction THIAMIN-PYROPHOSPHOKINASE-RXN at 0x13c396df0>,
  <Reaction THIAMINASE-RXN at 0x13c396f70>],
 [<Reaction ABC-32-RXN at 0x1294e6f40>, <Reaction RXNQT-4191 at 0x13c36e760>])

In [347]:
get_flux("RXNQT-4191")
get_metabolite_reactions("THIAMINE-P[c]")
get_flux("THI-P-SYN-RXN")
get_flux("RXN-12610")
get_flux("RXN-12611")

Growth rate: 0.5680 1/hr
Flux through RXNQT-4191: 0.0
Consuming reactions:
	RXNQT-4191: THIAMINE-P[c] + WATER[c] --> Pi[c] + THIAMINE[c]
Producing reactions:
	THI-P-SYN-RXN: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] + PROTON[c] + THZ-P[c] --> PPI[c] + THIAMINE-P[c]
	RXN-12610: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] + CPD-13576[c] + 2.0 PROTON[c] --> CARBON-DIOXIDE[c] + PPI[c] + THIAMINE-P[c]
	RXN-12611: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] + CPD-13575[c] + 2.0 PROTON[c] --> CARBON-DIOXIDE[c] + PPI[c] + THIAMINE-P[c]

Growth rate: 0.5680 1/hr
Flux through THI-P-SYN-RXN: 0.0
Growth rate: 0.5680 1/hr
Flux through RXN-12610: 0.0
Growth rate: 0.5680 1/hr
Flux through RXN-12611: 0.0


np.float64(0.0)

In [364]:
def add_sources(rxn_id, metabolite_ids):
    essential_rxn = model.reactions.get_by_id(rxn_id)
    with model:
        model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)
        essential_rxn.knock_out()
        
        for metabolite_id in metabolite_ids:
            metabolite = model.metabolites.get_by_id(metabolite_id)
            source = model.add_boundary(
                model.metabolites.get_by_id(metabolite_id),
                type="sink",
                reaction_id=metabolite_id + "_source",
                lb=-1 * baseline_flux,  # Set to exactly the lost flux from above!
                ub=0
            )
            print(f"New reaction: {source.id}\n\t{source.reaction}\t\t{source.bounds}")
        sol = model.optimize()
        print(f"Growth rate with new reaction: {sol.objective_value:.2f} 1/hr")
        print()
            
            
            
        sol = model.optimize()
        print(f"Growth rate {sol.objective_value:.4f} 1/hr")

add_sources("ABC-32-RXN", ["AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]", "THZ-P[c]"])
add_sources("ABC-32-RXN", ["AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]", "CPD-13576[c]"])
add_sources("ABC-32-RXN", ["AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]", "CPD-13575[c]"])

New reaction: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]_source
	AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] <-- 		(np.float64(-0.00012665498825421942), 0)
New reaction: THZ-P[c]_source
	THZ-P[c] <-- 		(np.float64(-0.00012665498825421942), 0)
Growth rate with new reaction: 0.57 1/hr

Growth rate 0.5680 1/hr
New reaction: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]_source
	AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] <-- 		(np.float64(-0.00012665498825421942), 0)
New reaction: CPD-13576[c]_source
	CPD-13576[c] <-- 		(np.float64(-0.00012665498825421942), 0)
Growth rate with new reaction: 0.57 1/hr

Growth rate 0.5680 1/hr
New reaction: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]_source
	AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] <-- 		(np.float64(-0.00012665498825421942), 0)
New reaction: CPD-13575[c]_source
	CPD-13575[c] <-- 		(np.float64(-0.00012665498825421942), 0)
Growth rate with new reaction: 0.57 1/hr

Growth rate 0.5680 1/hr


We are missing some combination of 

AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]", "THZ-P[c]

and/or 

AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]", "CPD-13576[c]

and/or 

AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]", "CPD-13575[c]

None of these metabolites are being produced. 

Let's first investigate AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]

In [369]:
get_metabolite_reactions("AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]")
add_sources("ABC-32-RXN", ["AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]", "THZ-P[c]"])
get_metabolite_reactions("AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]")
add_sources("ABC-32-RXN", ["AMINO-HYDROXYMETHYL-METHYL-PYR-P[c]", "THZ-P[c]"])
get_metabolite_reactions("AMINO-HYDROXYMETHYL-METHYL-PYR-P[c]")
add_sources("ABC-32-RXN", ["HMP[c]", "THZ-P[c]"])
get_metabolite_reactions("HMP[c]")

Consuming reactions:
	THI-P-SYN-RXN: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] + PROTON[c] + THZ-P[c] --> PPI[c] + THIAMINE-P[c]
	RXN-12610: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] + CPD-13576[c] + 2.0 PROTON[c] --> CARBON-DIOXIDE[c] + PPI[c] + THIAMINE-P[c]
	RXN-12611: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] + CPD-13575[c] + 2.0 PROTON[c] --> CARBON-DIOXIDE[c] + PPI[c] + THIAMINE-P[c]
Producing reactions:
	PYRIMSYN3-RXN: AMINO-HYDROXYMETHYL-METHYL-PYR-P[c] + ATP[c] --> ADP[c] + AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]

New reaction: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c]_source
	AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] <-- 		(np.float64(-0.00012665498825421942), 0)
New reaction: THZ-P[c]_source
	THZ-P[c] <-- 		(np.float64(-0.00012665498825421942), 0)
Growth rate with new reaction: 0.57 1/hr

Growth rate 0.5680 1/hr
Consuming reactions:
	THI-P-SYN-RXN: AMINO-HYDROXYMETHYL-METHYLPYRIMIDINE-PP[c] + PROTON[c] + THZ-P[c] --> PPI[c] + THIAMINE-P[c]
	RXN-12610: AMINO-HYDROXYME

([<Reaction OHMETPYRKIN-RXN at 0x12b396b80>],
 [<Reaction RXN-12613 at 0x13c0a4f40>,
  <Reaction THIAMINASE-RXN at 0x13c396f70>])

In [ ]:
m = model.metabolites.get_by_id("CPD-13578[c]")
print(m.id, m.name, m.formula)
get_metabolite_reactions("CPD-13578[c]")

m = model.metabolites.get_by_id("CPD-13577[c]")
print(m.id, m.name, m.formula)
get_metabolite_reactions("CPD-13577[c]")

CPD-13578[c] aminomethylpyrimidine C6H11N4
Consuming reactions:
	RXN-12613: CPD-13578[c] + WATER[c] --> AMMONIUM[c] + HMP[c]
Producing reactions:
	RXN-12612: CPD-13577[c] + WATER[c] --> CPD-13578[c] + FORMATE[c]

CPD-13577[c] formylaminopyrimidine C7H10N4O
Consuming reactions:
	RXN-12612: CPD-13577[c] + WATER[c] --> CPD-13578[c] + FORMATE[c]
Producing reactions:



([<Reaction RXN-12612 at 0x13c0bea90>], [])

formylaminopyrimidine has no known producing pathway... And yet must be required for thiamine synthesis... 

Note HMP = 4-amino-2-methyl-5-pyrimidinemethanol 

Note FAMP = N-formyl-4-amino-5-aminomethyl-2-methylpyrimidine

Note AmMP = 4-amino-5-aminomethyl-2-methylpyrimidine = CPD-13578[c]

However this paper (https://pmc.ncbi.nlm.nih.gov/articles/PMC10470507/) notes that R. pomeroyi DSS-3 grow only upon exogenously supplied AmMP and HMP, not FAMP (see Figure 4).

This implies that HMP and CPD-13578[c] must be transported into the cell somehow. 

In biocyc, we only see the following HMP reactions:

1. aminomethylpyrimidine + H2O → 4-amino-2-methyl-5-pyrimidinemethanol + ammonium	
2. thiamine + H2O → 5-(2-hydroxyethyl)-4-methylthiazole + 4-amino-2-methyl-5-pyrimidinemethanol + H+	
3. pyrithiamine + oxidized electron carrier + H2O = 4-amino-2-methyl-5-pyrimidinemethanol + 2-(2-methylpyridin-3-yl)ethanol + reduced two electron carrier

In biocyc, we only see the following CPD-13578 producing reactions:

1. formylaminopyrimidine + H2O → aminomethylpyrimidine + formate

Bref, we don't see any known HMP or CPD-13578 transporters, but it seems likely they exist. 

Moreover, the paper indicates that R Pom grows in B1 media (same figure)... It seems very likely that there is some kind of B1 transport going on. 

Moreover, ABC transporter permease [Ruegeria pomeroyi] via BLAST shows 38% sequence identity with E Coli B1 ABC transport. 

Probably best to keep ABC 32 RXN as it models B1 ABC transport, and consider this as an as-of-yet unidentified pathway.  


## 14. CHORPYRLY-RXN

In [379]:
rxn_id = "CHORPYRLY-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)



CHORPYRLY-RXN:
	CHORISMATE[c] --> 4-hydroxybenzoate[c] + PYRUVATE[c]

Growth rate with CHORPYRLY-RXN: 0.5680 1/hr
Flux through CHORPYRLY-RXN: 0.0001266549882542172

Growth rate without CHORPYRLY-RXN: 0.0000 1/hr


In [469]:
get_metabolite_reactions("4-hydroxybenzoate[c]")

Consuming reactions:
	2.5.1.39-RXN: 4-hydroxybenzoate[c] + CPD-21340[c] --> NONAPRENYL-4-HYDROXYBENZOATE[c] + PPI[c]
	4OHBENZOATE-OCTAPRENYLTRANSFER-RXN: 4-hydroxybenzoate[c] + OCTAPRENYL-DIPHOSPHATE[c] --> 3-OCTAPRENYL-4-HYDROXYBENZOATE[c] + PPI[c]
	RXN-9003: 4-hydroxybenzoate[c] + ALL-TRANS-HEXAPRENYL-DIPHOSPHATE[c] --> 3-HEXAPRENYL-4-HYDROXYBENZOATE[c] + PPI[c]
	RXN-9222: 4-hydroxybenzoate[c] + ALL-TRANS-HEPTAPRENYL-DIPHOSPHATE[c] --> CPD-9852[c] + PPI[c]
	RXN-9230: 4-hydroxybenzoate[c] + CPD-9610[c] --> CPD-9864[c] + PPI[c]
	4-HYDROXYBENZOATE-3-MONOOXYGENASE-RXN: 4-hydroxybenzoate[c] + NADPH[c] + OXYGEN-MOLECULE[c] + PROTON[c] --> 3-4-DIHYDROXYBENZOATE[c] + NADP[c] + WATER[c]
Producing reactions:
	CHORPYRLY-RXN: CHORISMATE[c] --> 4-hydroxybenzoate[c] + PYRUVATE[c]



([<Reaction 2.5.1.39-RXN at 0x13c71de80>,
  <Reaction 4OHBENZOATE-OCTAPRENYLTRANSFER-RXN at 0x12952e640>,
  <Reaction RXN-9003 at 0x13c706d60>,
  <Reaction RXN-9222 at 0x13c576d60>,
  <Reaction RXN-9230 at 0x13c7b6ac0>,
  <Reaction 4-HYDROXYBENZOATE-3-MONOOXYGENASE-RXN at 0x12931eaf0>],
 [<Reaction CHORPYRLY-RXN at 0x129674970>])

In [384]:
add_sink(rxn_id, "CHORISMATE[c]")
add_source(rxn_id, "PYRUVATE[c]")
add_source(rxn_id, "4-hydroxybenzoate[c]")


Growth rate with CHORPYRLY-RXN: 0.5680 1/hr
Flux through CHORPYRLY-RXN: 0.00012665498825421837

Growth rate without CHORPYRLY-RXN: 0.0000 1/hr
New reaction: CHORISMATE[c]_sink
	CHORISMATE[c] --> 		(0, np.float64(0.00012665498825421837))
Growth rate with new reaction: 0.00 1/hr

Growth rate with CHORPYRLY-RXN: 0.5680 1/hr
Flux through CHORPYRLY-RXN: 0.00012665498825421837

Growth rate without CHORPYRLY-RXN: 0.0000 1/hr
New reaction: PYRUVATE[c]_source
	PYRUVATE[c] <-- 		(np.float64(-0.00012665498825421837), 0)
Growth rate with new reaction: 0.00 1/hr

Growth rate with CHORPYRLY-RXN: 0.5680 1/hr
Flux through CHORPYRLY-RXN: 0.00012665498825421837

Growth rate without CHORPYRLY-RXN: 0.0000 1/hr
New reaction: 4-hydroxybenzoate[c]_source
	4-hydroxybenzoate[c] <-- 		(np.float64(-0.00012665498825421837), 0)
Growth rate with new reaction: 0.57 1/hr



In [387]:
get_metabolite_reactions("4-hydroxybenzoate[c]")
get_metabolite_reactions("CHORISMATE[c]")


Consuming reactions:
	4OHBENZOATE-OCTAPRENYLTRANSFER-RXN: 4-hydroxybenzoate[c] + OCTAPRENYL-DIPHOSPHATE[c] --> 3-OCTAPRENYL-4-HYDROXYBENZOATE[c] + PPI[c]
	RXN-9003: 4-hydroxybenzoate[c] + ALL-TRANS-HEXAPRENYL-DIPHOSPHATE[c] --> 3-HEXAPRENYL-4-HYDROXYBENZOATE[c] + PPI[c]
	2.5.1.39-RXN: 4-hydroxybenzoate[c] + CPD-21340[c] --> NONAPRENYL-4-HYDROXYBENZOATE[c] + PPI[c]
	RXN-9222: 4-hydroxybenzoate[c] + ALL-TRANS-HEPTAPRENYL-DIPHOSPHATE[c] --> CPD-9852[c] + PPI[c]
	RXN-9230: 4-hydroxybenzoate[c] + CPD-9610[c] --> CPD-9864[c] + PPI[c]
	4-HYDROXYBENZOATE-3-MONOOXYGENASE-RXN: 4-hydroxybenzoate[c] + NADPH[c] + OXYGEN-MOLECULE[c] + PROTON[c] --> 3-4-DIHYDROXYBENZOATE[c] + NADP[c] + WATER[c]
Producing reactions:
	CHORPYRLY-RXN: CHORISMATE[c] --> 4-hydroxybenzoate[c] + PYRUVATE[c]

Consuming reactions:
	CHORISMATEMUT-RXN: CHORISMATE[c] --> PREPHENATE[c]
	CHORPYRLY-RXN: CHORISMATE[c] --> 4-hydroxybenzoate[c] + PYRUVATE[c]
	PABASYN-RXN: CHORISMATE[c] + GLN[c] --> 4-AMINO-4-DEOXYCHORISMATE[c] + GLT[c]

([<Reaction CHORISMATEMUT-RXN at 0x129674820>,
  <Reaction CHORPYRLY-RXN at 0x129674970>,
  <Reaction PABASYN-RXN at 0x12b3a1790>,
  <Reaction ANTHRANSYN-RXN at 0x12964a5e0>],
 [<Reaction CHORISMATE-SYNTHASE-RXN at 0x1296747c0>])

This reaction is required for 4-hydroxybenzoate production, which is required for ubiquinol synthesis. 

However, this reaction is not in the biocyc database. Moreover, a BLAST search for this protein based on the E Coli variant returns no hits in R Pom.

4-hydroxybenzoate production seems to be tied to the following reactions:

1. dimethylsulfoniopropanoate + an acyl-CoA → dimethylsulfoniopropioyl-CoA + a carboxylate	(2.3.1.-)

2. an acyl-CoA + H2O → coenzyme A + a carboxylate + H+	(3.1.2.20)

3. an aryl sulfate + H2O → a phenol + sulfate + H+ (3.1.6.1)

In [414]:
search_metabolites(["dimethylsulfoniopropanoate"])

search_metabolites(["4-hydroxybenzoyl"])

search_metabolites(["4-hydroxybenzoate"])
search_metabolites(["dimethylsulfoniopropioyl"])

search_metabolites(["hydroxybenzoate"])


id                               | name                       | formula 
---------------------------------+----------------------------+---------
SS-DIMETHYL-BETA-PROPIOTHETIN[c] | dimethylsulfoniopropanoate | C5H10O2S
SS-DIMETHYL-BETA-PROPIOTHETIN[e] | dimethylsulfoniopropanoate | C5H10O2S
SS-DIMETHYL-BETA-PROPIOTHETIN[p] | dimethylsulfoniopropanoate | C5H10O2S
id         | name                 | formula       
-----------+----------------------+---------------
CPD-201[c] | 4-hydroxybenzoyl-CoA | C28H36N7O18P3S
id                                | name                            | formula 
----------------------------------+---------------------------------+---------
3-OCTAPRENYL-4-HYDROXYBENZOATE[c] | 3-octaprenyl-4-hydroxybenzoate  | C47H69O3
4-hydroxybenzoate[c]              | 4-hydroxybenzoate               | C7H5O3  
CPD-9852[c]                       | 3-heptaprenyl-4-hydroxybenzoate | C42H61O3
3-HEXAPRENYL-4-HYDROXYBENZOATE[c] | 3-hexaprenyl-4-hydroxybenzoate  | C37H53O3
NONAPREN

We have the metabolites in the model for the first 2 reaction types

1. dimethylsulfoniopropanoate + 4-hydroxybenzoyl-CoA → dimethylsulfoniopropioyl-CoA + 4-hydroxybenzoate

2. 4-hydroxybenzoyl-CoA + H2O → coenzyme A + 4-hydroxybenzoate + H+	

The third reaction would require 4-hydroxybenzoate-O-sulfate.

Naive implementation of these reactions fails due to lack of production of CPD-201 HBCOA. This is not produced because of a lack of CPD-10600 (3-(4-hydroxyphenyl)-3-oxo-propanoyl-CoA).

Unfortunately, CPD-10600 requires CPD-12199[c] (3S-(4-hydroxyphenyl)-3-hydroxy-propanoyl-CoA) which derives from P-COUMAROYL-COA[c].

P-COUMAROYL-COA[c] derives from either COUMARATE[c] or CPD-12199[c] (reverse reaction, so circular), neither of which have generating reactions in the model.

COUMARATE[c] can be generated by the following reactions:

1. dimethylsulfoniopropanoate + an acyl-CoA → dimethylsulfoniopropioyl-CoA + a carboxylate	(2.3.1.-)

2. an acyl-CoA + H2O → coenzyme A + a carboxylate + H+ (3.1.2.20)

The first reaction is circular with the downstream products. 

The second reaction would require (E)-p-coumaroyl-CoA, so this is also a circular reaction. 

______

we need to find another way of generating (E)-4-coumaroyl-CoA

From biocyc:

4-coumarate + coenzyme A + ATP → (E)-4-coumaroyl-CoA + AMP + diphosphate	

a (3Z)-3-enoyl-CoA → a (2E)-2-enoyl-CoA	fatty acid β-oxidation I, fatty acid β-oxidation III (unsaturated, odd number)

a 2,3,4-saturated fatty acyl CoA + oxidized electron-transfer flavoprotein + H+ → a (2E)-2-enoyl-CoA + reduced electron-transfer flavoprotein    fatty acid β-oxidation I

a (2E)-2-enoyl-CoA + NADP+ ← a (2E,4E)-2,4-dienoyl-CoA + NADPH + H+	

a [protein] S-acyl-L-cysteine + coenzyme A → a [protein]-L-cysteine + an acyl-CoA	

an acyl-CoA + acetyl-CoA ← a 3-oxoacyl-CoA + coenzyme A	

We could use:

2) 3-(4-hydroxyphenyl)propanoyl-CoA + ETF(ox) + H⁺ → (E)-4-coumaroyl-CoA + ETF(red) (NOT OK due to circular reasoning with 12199)

In [457]:
search_metabolites(["dienoyl"])

id                                    | name                                                            | formula        
--------------------------------------+-----------------------------------------------------------------+----------------
T2-C4-DECADIENYL-COA[c]               | trans-Delta2, cis-Delta4-decadienoyl-CoA                        | C31H46N7O17P3S 
Trans-D2-cis-D7-tetradecenoyl-ACPs[c] | a (2E,7Z)-tetradeca-2,7-dienoyl-[acp]                           | C14H23OX       
Trans-D3-cis-D9-hexadecenoyl-ACPs[c]  | a (2E,9Z)-hexadeca-2,9-dienoyl-[acp]                            | C16H27OX       
CPD-17813[c]                          | (2E,11Z)-hexadec-2,11-dienoyl-CoA                               | C37H58N7O17P3S 
cis-vaccen-2-enoyl-ACPs[c]            | a (2E,11Z)-octadeca-2,11-dienoyl-[acp]                          | C18H31OX       
Trans-D3-cis-D5-dodecenoyl-ACPs[c]    | a (3E,5Z)-dodeca-3,5-dienoyl-[acp]                              | C12H19OX       
CPD0-1158[c]            

In [452]:
get_metabolite_reactions("CPD-201[c]")

get_metabolite_reactions("CPD-10600[c]")

get_metabolite_reactions("CPD-12199[c]")

get_metabolite_reactions("P-COUMAROYL-COA[c]")

get_metabolite_reactions("COUMARATE[c]")

get_metabolite_reactions("CPD-12199[c]")

m = model.metabolites.get_by_id("P-COUMAROYL-COA[c]")
print(m.id, m.name, m.formula)

m = model.metabolites.get_by_id("CPD-12199[c]")
print(m.id, m.name, m.formula)


Consuming reactions:
	RXN-11246: ACETYL-COA[c] + CPD-201[c] <-- CO-A[c] + CPD-10600[c]
Producing reactions:

Consuming reactions:
Producing reactions:
	RXN-11245: CPD-12199[c] + NAD[c] <=> CPD-10600[c] + NADH[c] + PROTON[c]
	RXN-11246: ACETYL-COA[c] + CPD-201[c] <-- CO-A[c] + CPD-10600[c]

Consuming reactions:
	RXN-11245: CPD-12199[c] + NAD[c] <=> CPD-10600[c] + NADH[c] + PROTON[c]
	RXN-11244: CPD-12199[c] <=> P-COUMAROYL-COA[c] + WATER[c]
Producing reactions:

Consuming reactions:
Producing reactions:
	4-COUMARATE--COA-LIGASE-RXN: ATP[c] + CO-A[c] + COUMARATE[c] --> AMP[c] + P-COUMAROYL-COA[c] + PPI[c]
	RXN-11244: CPD-12199[c] <=> P-COUMAROYL-COA[c] + WATER[c]

Consuming reactions:
	4-COUMARATE--COA-LIGASE-RXN: ATP[c] + CO-A[c] + COUMARATE[c] --> AMP[c] + P-COUMAROYL-COA[c] + PPI[c]
Producing reactions:

Consuming reactions:
	RXN-11245: CPD-12199[c] + NAD[c] <=> CPD-10600[c] + NADH[c] + PROTON[c]
	RXN-11244: CPD-12199[c] <=> P-COUMAROYL-COA[c] + WATER[c]
Producing reactions:

P-COUMAR

In [451]:
search_metabolites(["coumaroyl"])

m = model.metabolites.get_by_id("CPD-12199[c]")
print(m.id, m.name, m.formula)





id                 | name                | formula       
-------------------+---------------------+---------------
P-COUMAROYL-COA[c] | (E)-4-coumaroyl-CoA | C30H38N7O18P3S
CPD-12199[c] 3S-(4-hydroxyphenyl)-3-hydroxy-propanoyl-CoA C30H40N7O19P3S


In [450]:
with model:
    model.reactions.get_by_id("CHORPYRLY-RXN").knock_out()
    
    # define relevant metabolites
    DMSP = model.metabolites.get_by_id("SS-DIMETHYL-BETA-PROPIOTHETIN[c]")
    HBCOA = model.metabolites.get_by_id("CPD-201[c]")
    DMSPCOA = model.metabolites.get_by_id("CPD-10473[c]")
    HB = model.metabolites.get_by_id("4-hydroxybenzoate[c]")
    COA = model.metabolites.get_by_id("CO-A[c]")
    PROTON = model.metabolites.get_by_id("PROTON[c]")
    WATER = model.metabolites.get_by_id("WATER[c]")
    
    # define reaction 2.3.1.-
    rxn1 = cobra.Reaction("RXN_2.3.1.-")
    rxn1.add_metabolites({DMSP: -1, HBCOA: -1, DMSPCOA: 1, HB: 1})
    model.add_reactions([rxn1])
    
    # define reaction 3.1.2.20
    rxn2 = cobra.Reaction("RXN_3.1.2.20")
    rxn2.add_metabolites({WATER: -1, HBCOA: -1, COA: 1, HB: 1, PROTON: 1})
    model.add_reactions([rxn2])
    
    metabolite_id = "P-COUMAROYL-COA[c]"
    source = model.add_boundary(
        model.metabolites.get_by_id(metabolite_id),
        type="sink",
        reaction_id=metabolite_id + "_source",
        lb=-1 * 0.00012665498825421837,  # Set to exactly the lost flux from above!
        ub=0
    )
    
    # test model
    model.reactions.get_by_id("EX_glc").bounds = (-5.44, 0)
    sol = model.optimize()
    print(f"Growth rate with {sol.objective_value:.4f} 1/hr")

Growth rate with 0.5680 1/hr


In [ ]:
coa = model.metabolites.get_by_id("CO-A[c]")

Mystery not solved. unsure.

## 15. ERYTH4PDEHYDROG-RXN

In [460]:
rxn_id = "ERYTH4PDEHYDROG-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)


add_sink(rxn_id, "ERYTHROSE-4P[c]")
print("---")
add_source(rxn_id, "ERYTHRONATE-4P[c]")




ERYTH4PDEHYDROG-RXN:
	ERYTHROSE-4P[c] + NAD[c] + WATER[c] --> ERYTHRONATE-4P[c] + NADH[c] + 2.0 PROTON[c]

Growth rate with ERYTH4PDEHYDROG-RXN: 0.5680 1/hr
Flux through ERYTH4PDEHYDROG-RXN: 0.00012665498825422278

Growth rate without ERYTH4PDEHYDROG-RXN: 0.0000 1/hr
Growth rate with ERYTH4PDEHYDROG-RXN: 0.5680 1/hr
Flux through ERYTH4PDEHYDROG-RXN: 0.00012665498825421652

Growth rate without ERYTH4PDEHYDROG-RXN: 0.0000 1/hr
New reaction: ERYTHROSE-4P[c]_sink
	ERYTHROSE-4P[c] --> 		(0, np.float64(0.00012665498825421652))
Growth rate with new reaction: 0.00 1/hr

---
Growth rate with ERYTH4PDEHYDROG-RXN: 0.5680 1/hr
Flux through ERYTH4PDEHYDROG-RXN: 0.00012665498825421666

Growth rate without ERYTH4PDEHYDROG-RXN: 0.0000 1/hr
New reaction: ERYTHRONATE-4P[c]_source
	ERYTHRONATE-4P[c] <-- 		(np.float64(-0.00012665498825421666), 0)
Growth rate with new reaction: 0.57 1/hr



In [462]:
m = model.metabolites.get_by_id("ERYTHROSE-4P[c]")
print(m.id, m.name, m.formula)

m = model.metabolites.get_by_id("ERYTHRONATE-4P[c]")
print(m.id, m.name, m.formula)


ERYTHROSE-4P[c] D-erythrose 4-phosphate C4H7O7P
ERYTHRONATE-4P[c] 4-phospho-D-erythronate C4H6O8P


Required for ERYTHRONATE-4P synthesis. 

This is in the biocyc database as reaction 1.2.1.72, the sole reaction which generates ERYTHRONATE-4P. 

Erythronate 4P is a required linear step in pyridoxal 5'-phosphate biosynthesis I explaining why it is required.

## 16. RXN-8992 

In [464]:
rxn_id = "RXN-8992"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)


RXN-8992:
	5.0 DELTA3-ISOPENTENYL-PP[c] + FARNESYL-PP[c] --> OCTAPRENYL-DIPHOSPHATE[c] + 5.0 PPI[c]

Growth rate with RXN-8992: 0.5680 1/hr
Flux through RXN-8992: 0.0001266549882542179

Growth rate without RXN-8992: 0.0000 1/hr


In [465]:
add_sink(rxn_id, "DELTA3-ISOPENTENYL-PP[c]")
add_sink(rxn_id, "FARNESYL-PP[c]")
add_source(rxn_id, "OCTAPRENYL-DIPHOSPHATE[c]")

Growth rate with RXN-8992: 0.5680 1/hr
Flux through RXN-8992: 0.00012665498825421758

Growth rate without RXN-8992: 0.0000 1/hr
New reaction: DELTA3-ISOPENTENYL-PP[c]_sink
	DELTA3-ISOPENTENYL-PP[c] --> 		(0, np.float64(0.00012665498825421758))
Growth rate with new reaction: 0.00 1/hr

Growth rate with RXN-8992: 0.5680 1/hr
Flux through RXN-8992: 0.00012665498825421758

Growth rate without RXN-8992: 0.0000 1/hr
New reaction: FARNESYL-PP[c]_sink
	FARNESYL-PP[c] --> 		(0, np.float64(0.00012665498825421758))
Growth rate with new reaction: 0.00 1/hr

Growth rate with RXN-8992: 0.5680 1/hr
Flux through RXN-8992: 0.00012665498825421758

Growth rate without RXN-8992: 0.0000 1/hr
New reaction: OCTAPRENYL-DIPHOSPHATE[c]_source
	OCTAPRENYL-DIPHOSPHATE[c] <-- 		(np.float64(-0.00012665498825421758), 0)
Growth rate with new reaction: 0.57 1/hr



Required for OCTAPRENYL-DIPHOSPHATE[c] (all-trans-octaprenyl diphosphate) production. OCTAPRENYL-DIPHOSPHATE[c] is required for ubiquinol-8 biosynthesis.

No known reaction for all-trans-octaprenyl diphosphate generation, RXN 8992 is not in biocyc database. 

The enzyme behind RXN 8992 is Octaprenyl diphosphate synthase, and if we BLAST the E Coli variant of this enzyme, we are able to get 32-47% hits with 	a polyprenyl synthetase family protein in Ruegeria alba and marina cluster representative sequences under Ruegeria pomeroyi records in the NIH BLAST program. 

This suggests that this reaction could be happening nonetheless, with a unique variant of Octaprenyl diphosphate synthase.


In [466]:
m = model.metabolites.get_by_id("OCTAPRENYL-DIPHOSPHATE[c]")
print(m.id, m.name, m.formula)

get_metabolite_reactions("OCTAPRENYL-DIPHOSPHATE[c]")




OCTAPRENYL-DIPHOSPHATE[c] all-trans-octaprenyl diphosphate C40H65O7P2
Consuming reactions:
	4OHBENZOATE-OCTAPRENYLTRANSFER-RXN: 4-hydroxybenzoate[c] + OCTAPRENYL-DIPHOSPHATE[c] --> 3-OCTAPRENYL-4-HYDROXYBENZOATE[c] + PPI[c]
Producing reactions:
	RXN-8992: 5.0 DELTA3-ISOPENTENYL-PP[c] + FARNESYL-PP[c] --> OCTAPRENYL-DIPHOSPHATE[c] + 5.0 PPI[c]



([<Reaction 4OHBENZOATE-OCTAPRENYLTRANSFER-RXN at 0x12952e640>],
 [<Reaction RXN-8992 at 0x13c266d60>])

## 17. DIHYDROFOLATEREDUCT-RXN

In [467]:
rxn_id = "DIHYDROFOLATEREDUCT-RXN"
essential_rxn = model.reactions.get_by_id(rxn_id)
print(f"{essential_rxn.id}:\n\t{essential_rxn.reaction}\n")

test_essential, baseline_flux = check_essential_rxn(rxn_id)



DIHYDROFOLATEREDUCT-RXN:
	DIHYDROFOLATE-GLU-N[c] + NADPH[c] + PROTON[c] --> NADP[c] + THF-GLU-N[c]

Growth rate with DIHYDROFOLATEREDUCT-RXN: 0.5680 1/hr
Flux through DIHYDROFOLATEREDUCT-RXN: 0.015241196315316981

Growth rate without DIHYDROFOLATEREDUCT-RXN: 0.5680 1/hr


Looks non-essential.

## Ubiquinol and Ubiquninone investigation

https://pubmed.ncbi.nlm.nih.gov/20228209/ notes that Ubiquinone 10 is the primary ubiquinone of R Pom. 

Maybe we can eliminate the Ubquinol 6 and 10 we have in the model.

In [470]:
search_metabolites(["ubiqui"])

id                                    | name                         | formula 
--------------------------------------+------------------------------+---------
CPD-9958[c]                           | ubiquinol-10                 | C59H92O4
UBIQUINONE-10[c]                      | ubiquinone-10                | C59H90O4
UBIQUINOL-30[c]                       | ubiquinol-6                  | C39H60O4
UBIQUINONE-6[c]                       | ubiquinone-6                 | C39H58O4
CPD-9956[c]                           | ubiquinol-8                  | C49H76O4
OCTAPRENYL-METHYL-OH-METHOXY-BENZQ[c] | 3-demethylubiquinol-8        | C48H74O4
UBIQUINONE-8[c]                       | ubiquinone-8                 | C49H74O4
CPD-9873[c]                           | 3-demethylubiquinol-10       | C58H90O4
CPD-9861[c]                           | 3-demethylubiquinol-<i>7</i> | C43H66O4
CPD-9955[c]                           | ubiquinol-<i>7</i>           | C44H68O4
